<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Max Kuhn, y Julia Silge">
<title>Modelado Ordenado con R - 9&nbsp; Judging Model Effectiveness</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./10-resampling.html" rel="next">
<link href="./08-feature-engineering.html" rel="prev">
<link href="./images/cover.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><link rel="stylesheet" href="style.css">
<link rel="stylesheet" href="TMwR.css">
</head>
<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./04-ames.html">MODELING BASICS</a></li><li class="breadcrumb-item"><a href="./09-judging-model-effectiveness.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Judging Model Effectiveness</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Modelado Ordenado con R</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/davidrsch/TMwRes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hello World</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">INTRODUCTION</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-software-modeling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Software for modeling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">A Tidyverse Primer</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-base-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">A Review of R Modeling Fundamentals</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">MODELING BASICS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-ames.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">The Ames Housing Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-data-spending.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Spending our Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-fitting-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Fitting Models with parsnip</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-the-model-workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">A Model Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-feature-engineering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Feature Engineering with recipes</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-judging-model-effectiveness.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Judging Model Effectiveness</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">TOOLS FOR CREATING EFFECTIVE MODELS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-resampling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Resampling for Evaluating Performance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-comparing-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Comparing Models with Resampling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12-tuning-parameters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Model Tuning and the Dangers of Overfitting</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13-grid-search.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Grid Search</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14-iterative-search.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Iterative Search</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./15-workflow-sets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Screening Many Models</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">BEYOND THE BASICS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./16-dimensionality-reduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Dimensionality Reduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./17-encoding-categorical-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Encoding Categorical Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./18-explaining-models-and-predictions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Explaining Models and Predictions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./19-when-should-you-trust-predictions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">When Should You Trust Your Predictions?</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./20-ensemble-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Ensembles of Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21-inferential-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Inferential Analysis</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pre-proc-table.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Recommended Preprocessing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#performance-metrics-and-inference" id="toc-performance-metrics-and-inference" class="nav-link active" data-scroll-target="#performance-metrics-and-inference"><span class="header-section-number">9.1</span> Performance Metrics and Inference</a></li>
  <li><a href="#regression-metrics" id="toc-regression-metrics" class="nav-link" data-scroll-target="#regression-metrics"><span class="header-section-number">9.2</span> Regression Metrics</a></li>
  <li><a href="#binary-classification-metrics" id="toc-binary-classification-metrics" class="nav-link" data-scroll-target="#binary-classification-metrics"><span class="header-section-number">9.3</span> Binary Classification Metrics</a></li>
  <li><a href="#multiclass-classification-metrics" id="toc-multiclass-classification-metrics" class="nav-link" data-scroll-target="#multiclass-classification-metrics"><span class="header-section-number">9.4</span> Multiclass Classification Metrics</a></li>
  <li><a href="#sec-performance-summary" id="toc-sec-performance-summary" class="nav-link" data-scroll-target="#sec-performance-summary"><span class="header-section-number">9.5</span> Chapter Summary</a></li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/davidrsch/TMwRes/edit/main/09-judging-model-effectiveness.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/davidrsch/TMwRes/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-performance" class="quarto-section-identifier"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Judging Model Effectiveness</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p>Once we have a model, we need to know how well it works. A quantitative approach for estimating effectiveness allows us to understand the model, to compare different models, or to tweak the model to improve performance. Our focus in tidymodels is on empirical validation; this usually means using data that were not used to create the model as the substrate to measure effectiveness.</p>
<div class="rmdwarning">
<p>The best approach to empirical validation involves using <em>resampling</em> methods that will be introduced in <span class="quarto-unresolved-ref">?sec-resampling</span>. In this chapter, we will motivate the need for empirical validation by using the test set. Keep in mind that the test set can only be used once, as explained in <a href="05-data-spending.html#sec-splitting-methods"><span>Section&nbsp;5.1</span></a>.</p>
</div>
<p>When judging model effectiveness, your decision about which metrics to examine can be critical. In later chapters, certain model parameters will be empirically optimized and a primary performance metric will be used to choose the best sub-model. Choosing the wrong metric can easily result in unintended consequences. For example, two common metrics for regression models are the root mean squared error (RMSE) and the coefficient of determination (a.k.a. <span class="math inline">\(R^2\)</span>). The former measures <em>accuracy</em> while the latter measures <em>correlation</em>. These are not necessarily the same thing. <a href="#fig-performance-reg-metrics">Figure&nbsp;<span>9.1</span></a> demonstrates the difference between the two.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-performance-reg-metrics" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="figures/fig-performance-reg-metrics-1.png" class="img-fluid figure-img" alt="Scatter plots of numeric observed versus predicted values for models that are optimized using the RMSE and the coefficient of determination. The former results in results that are close to the 45 degree line of identity while the latter shows results with a tight linear correlation but falls well off of the line of identity." width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;9.1: Observed versus predicted values for models that are optimized using the RMSE compared to the coefficient of determination</figcaption></figure>
</div>
</div>
</div>
<p>A model optimized for RMSE has more variability but has relatively uniform accuracy across the range of the outcome. The right panel shows that there is a tighter correlation between the observed and predicted values but this model performs poorly in the tails.</p>
<p>This chapter will demonstrate the <span class="pkg">yardstick</span> package, a core tidymodels packages with the focus of measuring model performance. Before illustrating syntax, let’s explore whether empirical validation using performance metrics is worthwhile when a model is focused on inference rather than prediction.</p>
<section id="performance-metrics-and-inference" class="level2" data-number="9.1"><h2 data-number="9.1" class="anchored" data-anchor-id="performance-metrics-and-inference">
<span class="header-section-number">9.1</span> Performance Metrics and Inference</h2>
<p>The effectiveness of any given model depends on how the model will be used. An inferential model is used primarily to understand relationships, and typically emphasizes the choice (and validity) of probabilistic distributions and other generative qualities that define the model. For a model used primarily for prediction, by contrast, predictive strength is of primary importance and other concerns about underlying statistical qualities may be less important. Predictive strength is usually determined by how close our predictions come to the observed data, i.e., fidelity of the model predictions to the actual results. This chapter focuses on functions that can be used to measure predictive strength. However, our advice for those developing inferential models is to use these techniques even when the model will not be used with the primary goal of prediction.</p>
<p>A longstanding issue with the practice of inferential statistics is that, with a focus purely on inference, it is difficult to assess the credibility of a model. For example, consider the Alzheimer’s disease data from <span class="citation" data-cites="CraigSchapiro">Craig–Schapiro et al. (<a href="references.html#ref-CraigSchapiro" role="doc-biblioref">2011</a>)</span> when 333 patients were studied to determine the factors that influence cognitive impairment. An analysis might take the known risk factors and build a logistic regression model where the outcome is binary (impaired/non-impaired). Let’s consider predictors for age, sex, and the Apolipoprotein E genotype. The latter is a categorical variable with the six possible combinations of the three main variants of this gene. Apolipoprotein E is known to have an association with dementia <span class="citation" data-cites="Kim:2009p4370">(<a href="references.html#ref-Kim:2009p4370" role="doc-biblioref">Jungsu, Basak, and Holtzman 2009</a>)</span>.</p>
<p>A superficial, but not uncommon, approach to this analysis would be to fit a large model with main effects and interactions, then use statistical tests to find the minimal set of model terms that are statistically significant at some pre-defined level. If a full model with the three factors and their two- and three-way interactions were used, an initial phase would be to test the interactions using sequential likelihood ratio tests <span class="citation" data-cites="HosmerLemeshow">(<a href="references.html#ref-HosmerLemeshow" role="doc-biblioref">Hosmer and Lemeshow 2000</a>)</span>. Let’s step through this kind of approach for the example Alzheimer’s disease data:</p>
<ul>
<li><p>When comparing the model with all two-way interactions to one with the additional three-way interaction, the likelihood ratio tests produces a p-value of 0.888. This implies that there is no evidence that the four additional model terms associated with the three-way interaction explain enough of the variation in the data to keep them in the model.</p></li>
<li><p>Next, the two-way interactions are similarly evaluated against the model with no interactions. The p-value here is 0.0382. This is somewhat borderline, but, given the small sample size, it would be prudent to conclude that there is evidence that some of the 10 possible two-way interactions are important to the model.</p></li>
<li><p>From here, we would build some explanation of the results. The interactions would be particularly important to discuss since they may spark interesting physiological or neurological hypotheses to be explored further.</p></li>
</ul>
<p>While shallow, this analysis strategy is common in practice as well as in the literature. This is especially true if the practitioner has limited formal training in data analysis.</p>
<p>One missing piece of information in this approach is how closely this model fits the actual data. Using resampling methods, discussed in <span class="quarto-unresolved-ref">?sec-resampling</span>, we can estimate the accuracy of this model to be about 73%. Accuracy is often a poor measure of model performance; we use it here because it is commonly understood. If the model has 73% fidelity to the data, should we trust conclusions it produces? We might think so until we realize that the baseline rate of nonimpaired patients in the data is 72.7%. This means that, despite our statistical analysis, the two-factor model appears to be only 0.3% better than a simple heuristic that always predicts patients to be unimpaired, regardless of the observed data.</p>
<div class="rmdnote">
<p>The point of this analysis is to demonstrate the idea that optimization of statistical characteristics of the model does not imply that the model fits the data well. Even for purely inferential models, some measure of fidelity to the data should accompany the inferential results. Using this, the consumers of the analyses can calibrate their expectations of the results.</p>
</div>
<p>In the remainder of this chapter, we will discuss general approaches for evaluating models via empirical validation. These approaches are grouped by the nature of the outcome data: purely numeric, binary classes, and three or more class levels.</p>
</section><section id="regression-metrics" class="level2" data-number="9.2"><h2 data-number="9.2" class="anchored" data-anchor-id="regression-metrics">
<span class="header-section-number">9.2</span> Regression Metrics</h2>
<p>Recall from <a href="06-fitting-models.html#sec-parsnip-predictions"><span>Section&nbsp;6.3</span></a> that tidymodels prediction functions produce tibbles with columns for the predicted values. These columns have consistent names, and the functions in the <span class="pkg">yardstick</span> package that produce performance metrics have consistent interfaces. The functions are data frame-based, as opposed to vector-based, with the general syntax of:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">function</span>(data, truth, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>where <code>data</code> is a data frame or tibble and <code>truth</code> is the column with the observed outcome values. The ellipses or other arguments are used to specify the column(s) containing the predictions.</p>
<p>To illustrate, let’s take the model from <a href="08-feature-engineering.html#sec-recipes-summary"><span>Section&nbsp;8.8</span></a>. This model <code>lm_wflow_fit</code> combines a linear regression model with a predictor set supplemented with an interaction and spline functions for longitude and latitude. It was created from a training set (named <code>ames_train</code>). Although we do not advise using the test set at this juncture of the modeling process, it will be used here to illustrate functionality and syntax. The data frame <code>ames_test</code> consists of 588 properties. To start, let’s produce predictions:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ames_test_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">lm_fit</span>, new_data <span class="op">=</span> <span class="va">ames_test</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">Sale_Price</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ames_test_res</span></span>
<span><span class="co">## # A tibble: 588 × 1</span></span>
<span><span class="co">##   .pred</span></span>
<span><span class="co">##   &lt;dbl&gt;</span></span>
<span><span class="co">## 1  5.07</span></span>
<span><span class="co">## 2  5.31</span></span>
<span><span class="co">## 3  5.28</span></span>
<span><span class="co">## 4  5.33</span></span>
<span><span class="co">## 5  5.30</span></span>
<span><span class="co">## 6  5.24</span></span>
<span><span class="co">## # ℹ 582 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The predicted numeric outcome from the regression model is named <code>.pred</code>. Let’s match the predicted values with their corresponding observed outcome values:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ames_test_res</span> <span class="op">&lt;-</span> <span class="fu">bind_cols</span><span class="op">(</span><span class="va">ames_test_res</span>, <span class="va">ames_test</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">Sale_Price</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ames_test_res</span></span>
<span><span class="co">## # A tibble: 588 × 2</span></span>
<span><span class="co">##   .pred Sale_Price</span></span>
<span><span class="co">##   &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">## 1  5.07       5.02</span></span>
<span><span class="co">## 2  5.31       5.39</span></span>
<span><span class="co">## 3  5.28       5.28</span></span>
<span><span class="co">## 4  5.33       5.28</span></span>
<span><span class="co">## 5  5.30       5.28</span></span>
<span><span class="co">## 6  5.24       5.26</span></span>
<span><span class="co">## # ℹ 582 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We see that these values mostly look close, but we don’t yet have a quantitative understanding of how the model is doing because we haven’t computed any performance metrics. Note that both the predicted and observed outcomes are in log-10 units. It is best practice to analyze the predictions on the transformed scale (if one were used) even if the predictions are reported using the original units.</p>
<p>Let’s plot the data in <a href="#fig-ames-performance-plot">Figure&nbsp;<span>9.2</span></a> before computing metrics:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">ames_test_res</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sale_Price</span>, y <span class="op">=</span> <span class="va">.pred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="co"># Create a diagonal line:</span></span>
<span>  <span class="fu">geom_abline</span><span class="op">(</span>lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Predicted Sale Price (log10)"</span>, x <span class="op">=</span> <span class="st">"Sale Price (log10)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Scale and size the x- and y-axis uniformly:</span></span>
<span>  <span class="fu">coord_obs_pred</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-ames-performance-plot" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="figures/fig-ames-performance-plot-1.png" class="img-fluid figure-img" alt="Scatter plots of numeric observed versus predicted values for an Ames regression model. Both axes use log-10 units. The model shows good concordance with some poorly fitting points at high and low prices." width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;9.2: Observed versus predicted values for an Ames regression model, with log-10 units on both axes</figcaption></figure>
</div>
</div>
</div>
<p>There is one low-price property that is substantially over-predicted, i.e., quite high above the dashed line.</p>
<p>Let’s compute the root mean squared error for this model using the <code>rmse()</code> function:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">rmse</span><span class="op">(</span><span class="va">ames_test_res</span>, truth <span class="op">=</span> <span class="va">Sale_Price</span>, estimate <span class="op">=</span> <span class="va">.pred</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 rmse    standard      0.0736</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This shows us the standard format of the output of <span class="pkg">yardstick</span> functions. Metrics for numeric outcomes usually have a value of “standard” for the <code>.estimator</code> column. Examples with different values for this column are shown in the next sections.</p>
<p>To compute multiple metrics at once, we can create a <em>metric set</em>. Let’s add <span class="math inline">\(R^2\)</span> and the mean absolute error:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ames_metrics</span> <span class="op">&lt;-</span> <span class="fu">metric_set</span><span class="op">(</span><span class="va">rmse</span>, <span class="va">rsq</span>, <span class="va">mae</span><span class="op">)</span></span>
<span><span class="fu">ames_metrics</span><span class="op">(</span><span class="va">ames_test_res</span>, truth <span class="op">=</span> <span class="va">Sale_Price</span>, estimate <span class="op">=</span> <span class="va">.pred</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 3 × 3</span></span>
<span><span class="co">##   .metric .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 rmse    standard      0.0736</span></span>
<span><span class="co">## 2 rsq     standard      0.836 </span></span>
<span><span class="co">## 3 mae     standard      0.0549</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This tidy data format stacks the metrics vertically. The root mean squared error and mean absolute error metrics are both on the scale of the outcome (so <code>log10(Sale_Price)</code> for our example) and measure the difference between the predicted and observed values. The value for <span class="math inline">\(R^2\)</span> measures the squared correlation between the predicted and observed values, so values closer to one are better.</p>
<div class="rmdwarning">
<p>The <span class="pkg">yardstick</span> package does <em>not</em> contain a function for adjusted <span class="math inline">\(R^2\)</span>. This modification of the coefficient of determination is commonly used when the same data used to fit the model are used to evaluate the model. This metric is not fully supported in tidymodels because it is always a better approach to compute performance on a separate data set than the one used to fit the model.</p>
</div>
</section><section id="binary-classification-metrics" class="level2" data-number="9.3"><h2 data-number="9.3" class="anchored" data-anchor-id="binary-classification-metrics">
<span class="header-section-number">9.3</span> Binary Classification Metrics</h2>
<p>To illustrate other ways to measure model performance, we will switch to a different example. The <span class="pkg">modeldata</span> package (another one of the tidymodels packages) contains example predictions from a test data set with two classes (“Class1” and “Class2”):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">two_class_example</span><span class="op">)</span></span>
<span><span class="fu">tibble</span><span class="op">(</span><span class="va">two_class_example</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 500 × 4</span></span>
<span><span class="co">##   truth   Class1   Class2 predicted</span></span>
<span><span class="co">##   &lt;fct&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;fct&gt;    </span></span>
<span><span class="co">## 1 Class2 0.00359 0.996    Class2   </span></span>
<span><span class="co">## 2 Class1 0.679   0.321    Class1   </span></span>
<span><span class="co">## 3 Class2 0.111   0.889    Class2   </span></span>
<span><span class="co">## 4 Class1 0.735   0.265    Class1   </span></span>
<span><span class="co">## 5 Class2 0.0162  0.984    Class2   </span></span>
<span><span class="co">## 6 Class1 0.999   0.000725 Class1   </span></span>
<span><span class="co">## # ℹ 494 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The second and third columns are the predicted class probabilities for the test set while <code>predicted</code> are the discrete predictions.</p>
<p>For the hard class predictions, a variety of <span class="pkg">yardstick</span> functions are helpful:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># A confusion matrix: </span></span>
<span><span class="fu">conf_mat</span><span class="op">(</span><span class="va">two_class_example</span>, truth <span class="op">=</span> <span class="va">truth</span>, estimate <span class="op">=</span> <span class="va">predicted</span><span class="op">)</span></span>
<span><span class="co">##           Truth</span></span>
<span><span class="co">## Prediction Class1 Class2</span></span>
<span><span class="co">##     Class1    227     50</span></span>
<span><span class="co">##     Class2     31    192</span></span>
<span></span>
<span><span class="co"># Accuracy:</span></span>
<span><span class="fu">accuracy</span><span class="op">(</span><span class="va">two_class_example</span>, <span class="va">truth</span>, <span class="va">predicted</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric  .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;    &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 accuracy binary         0.838</span></span>
<span></span>
<span><span class="co"># Matthews correlation coefficient:</span></span>
<span><span class="fu">mcc</span><span class="op">(</span><span class="va">two_class_example</span>, <span class="va">truth</span>, <span class="va">predicted</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 mcc     binary         0.677</span></span>
<span></span>
<span><span class="co"># F1 metric:</span></span>
<span><span class="fu">f_meas</span><span class="op">(</span><span class="va">two_class_example</span>, <span class="va">truth</span>, <span class="va">predicted</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 f_meas  binary         0.849</span></span>
<span></span>
<span><span class="co"># Combining these three classification metrics together</span></span>
<span><span class="va">classification_metrics</span> <span class="op">&lt;-</span> <span class="fu">metric_set</span><span class="op">(</span><span class="va">accuracy</span>, <span class="va">mcc</span>, <span class="va">f_meas</span><span class="op">)</span></span>
<span><span class="fu">classification_metrics</span><span class="op">(</span><span class="va">two_class_example</span>, truth <span class="op">=</span> <span class="va">truth</span>, estimate <span class="op">=</span> <span class="va">predicted</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 3 × 3</span></span>
<span><span class="co">##   .metric  .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;    &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 accuracy binary         0.838</span></span>
<span><span class="co">## 2 mcc      binary         0.677</span></span>
<span><span class="co">## 3 f_meas   binary         0.849</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The Matthews correlation coefficient and F1 score both summarize the confusion matrix, but compared to <code>mcc()</code>, which measures the quality of both positive and negative examples, the <code>f_meas()</code> metric emphasizes the positive class, i.e., the event of interest. For binary classification data sets like this example, <span class="pkg">yardstick</span> functions have a standard argument called <code>event_level</code> to distinguish positive and negative levels. The default (which we used in this code) is that the <em>first</em> level of the outcome factor is the event of interest.</p>
<div class="rmdnote">
<p>There is some heterogeneity in R functions in this regard; some use the first level and others the second to denote the event of interest. We consider it more intuitive that the first level is the most important. The second level logic is borne of encoding the outcome as 0/1 (in which case the second value is the event) and unfortunately remains in some packages. However, tidymodels (along with many other R packages) require a categorical outcome to be encoded as a factor and, for this reason, the legacy justification for the second level as the event becomes irrelevant.</p>
</div>
<p>As an example where the second level is the event:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">f_meas</span><span class="op">(</span><span class="va">two_class_example</span>, <span class="va">truth</span>, <span class="va">predicted</span>, event_level <span class="op">=</span> <span class="st">"second"</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 f_meas  binary         0.826</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this output, the <code>.estimator</code> value of “binary” indicates that the standard formula for binary classes will be used.</p>
<p>There are numerous classification metrics that use the predicted probabilities as inputs rather than the hard class predictions. For example, the receiver operating characteristic (ROC) curve computes the sensitivity and specificity over a continuum of different event thresholds. The predicted class column is not used. There are two <span class="pkg">yardstick</span> functions for this method: <code>roc_curve()</code> computes the data points that make up the ROC curve and <code>roc_auc()</code> computes the area under the curve.</p>
<p>The interfaces to these types of metric functions use the <code>...</code> argument placeholder to pass in the appropriate class probability column. For two-class problems, the probability column for the event of interest is passed into the function:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">two_class_curve</span> <span class="op">&lt;-</span> <span class="fu">roc_curve</span><span class="op">(</span><span class="va">two_class_example</span>, <span class="va">truth</span>, <span class="va">Class1</span><span class="op">)</span></span>
<span><span class="va">two_class_curve</span></span>
<span><span class="co">## # A tibble: 502 × 3</span></span>
<span><span class="co">##   .threshold specificity sensitivity</span></span>
<span><span class="co">##        &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;</span></span>
<span><span class="co">## 1 -Inf           0                 1</span></span>
<span><span class="co">## 2    1.79e-7     0                 1</span></span>
<span><span class="co">## 3    4.50e-6     0.00413           1</span></span>
<span><span class="co">## 4    5.81e-6     0.00826           1</span></span>
<span><span class="co">## 5    5.92e-6     0.0124            1</span></span>
<span><span class="co">## 6    1.22e-5     0.0165            1</span></span>
<span><span class="co">## # ℹ 496 more rows</span></span>
<span></span>
<span><span class="fu">roc_auc</span><span class="op">(</span><span class="va">two_class_example</span>, <span class="va">truth</span>, <span class="va">Class1</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 roc_auc binary         0.939</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>two_class_curve</code> object can be used in a <code>ggplot</code> call to visualize the curve, as shown in <a href="#fig-example-roc-curve">Figure&nbsp;<span>9.3</span></a>. There is an <code>autoplot()</code> method that will take care of the details:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">autoplot</span><span class="op">(</span><span class="va">two_class_curve</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-example-roc-curve" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="figures/fig-example-roc-curve-1.png" class="img-fluid figure-img" alt="An example ROC curve. The x-axis is one minus specificity and the y-axis is sensitivity. The curve bows towards the upper left-hand side of the plot area." width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;9.3: Example ROC curve</figcaption></figure>
</div>
</div>
</div>
<p>If the curve was close to the diagonal line, then the model’s predictions would be no better than random guessing. Since the curve is up in the top, left-hand corner, we see that our model performs well at different thresholds.</p>
<p>There are a number of other functions that use probability estimates, including <code>gain_curve()</code>, <code>lift_curve()</code>, and <code>pr_curve()</code>.</p>
</section><section id="multiclass-classification-metrics" class="level2" data-number="9.4"><h2 data-number="9.4" class="anchored" data-anchor-id="multiclass-classification-metrics">
<span class="header-section-number">9.4</span> Multiclass Classification Metrics</h2>
<p>What about data with three or more classes? To demonstrate, let’s explore a different example data set that has four classes:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">hpc_cv</span><span class="op">)</span></span>
<span><span class="fu">tibble</span><span class="op">(</span><span class="va">hpc_cv</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 3,467 × 7</span></span>
<span><span class="co">##   obs   pred     VF      F       M          L Resample</span></span>
<span><span class="co">##   &lt;fct&gt; &lt;fct&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;   </span></span>
<span><span class="co">## 1 VF    VF    0.914 0.0779 0.00848 0.0000199  Fold01  </span></span>
<span><span class="co">## 2 VF    VF    0.938 0.0571 0.00482 0.0000101  Fold01  </span></span>
<span><span class="co">## 3 VF    VF    0.947 0.0495 0.00316 0.00000500 Fold01  </span></span>
<span><span class="co">## 4 VF    VF    0.929 0.0653 0.00579 0.0000156  Fold01  </span></span>
<span><span class="co">## 5 VF    VF    0.942 0.0543 0.00381 0.00000729 Fold01  </span></span>
<span><span class="co">## 6 VF    VF    0.951 0.0462 0.00272 0.00000384 Fold01  </span></span>
<span><span class="co">## # ℹ 3,461 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As before, there are factors for the observed and predicted outcomes along with four other columns of predicted probabilities for each class. (These data also include a <code>Resample</code> column. These <code>hpc_cv</code> results are for out-of-sample predictions associated with 10-fold cross-validation. For the time being, this column will be ignored and we’ll discuss resampling in depth in <span class="quarto-unresolved-ref">?sec-resampling</span>.)</p>
<p>The functions for metrics that use the discrete class predictions are identical to their binary counterparts:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">accuracy</span><span class="op">(</span><span class="va">hpc_cv</span>, <span class="va">obs</span>, <span class="va">pred</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric  .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;    &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 accuracy multiclass     0.709</span></span>
<span></span>
<span><span class="fu">mcc</span><span class="op">(</span><span class="va">hpc_cv</span>, <span class="va">obs</span>, <span class="va">pred</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 mcc     multiclass     0.515</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that, in these results, a “multiclass” <code>.estimator</code> is listed. Like “binary,” this indicates that the formula for outcomes with three or more class levels was used. The Matthews correlation coefficient was originally designed for two classes but has been extended to cases with more class levels.</p>
<p>There are methods for taking metrics designed to handle outcomes with only two classes and extend them for outcomes with more than two classes. For example, a metric such as sensitivity measures the true positive rate which, by definition, is specific to two classes (i.e., “event” and “nonevent”). How can this metric be used in our example data?</p>
<p>There are wrapper methods that can be used to apply sensitivity to our four-class outcome. These options are macro-averaging, macro-weighted averaging, and micro-averaging:</p>
<ul>
<li><p>Macro-averaging computes a set of one-versus-all metrics using the standard two-class statistics. These are averaged.</p></li>
<li><p>Macro-weighted averaging does the same but the average is weighted by the number of samples in each class.</p></li>
<li><p>Micro-averaging computes the contribution for each class, aggregates them, then computes a single metric from the aggregates.</p></li>
</ul>
<p>See <span class="citation" data-cites="wu2017unified">Wu and Zhou (<a href="references.html#ref-wu2017unified" role="doc-biblioref">2017</a>)</span> and <span class="citation" data-cites="OpitzBurst">Opitz and Burst (<a href="references.html#ref-OpitzBurst" role="doc-biblioref">2019</a>)</span> for more on extending classification metrics to outcomes with more than two classes.</p>
<p>Using sensitivity as an example, the usual two-class calculation is the ratio of the number of correctly predicted events divided by the number of true events. The manual calculations for these averaging methods are:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">class_totals</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">hpc_cv</span>, <span class="va">obs</span>, name <span class="op">=</span> <span class="st">"totals"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>class_wts <span class="op">=</span> <span class="va">totals</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">totals</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">class_totals</span></span>
<span><span class="co">##   obs totals class_wts</span></span>
<span><span class="co">## 1  VF   1769   0.51024</span></span>
<span><span class="co">## 2   F   1078   0.31093</span></span>
<span><span class="co">## 3   M    412   0.11883</span></span>
<span><span class="co">## 4   L    208   0.05999</span></span>
<span></span>
<span><span class="va">cell_counts</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">hpc_cv</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">obs</span>, <span class="va">pred</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute the four sensitivities using 1-vs-all</span></span>
<span><span class="va">one_versus_all</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">cell_counts</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">obs</span> <span class="op">==</span> <span class="va">pred</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">full_join</span><span class="op">(</span><span class="va">class_totals</span>, by <span class="op">=</span> <span class="st">"obs"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>sens <span class="op">=</span> <span class="va">n</span> <span class="op">/</span> <span class="va">totals</span><span class="op">)</span></span>
<span><span class="va">one_versus_all</span></span>
<span><span class="co">## # A tibble: 4 × 6</span></span>
<span><span class="co">##   obs   pred      n totals class_wts  sens</span></span>
<span><span class="co">##   &lt;fct&gt; &lt;fct&gt; &lt;int&gt;  &lt;int&gt;     &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">## 1 VF    VF     1620   1769    0.510  0.916</span></span>
<span><span class="co">## 2 F     F       647   1078    0.311  0.600</span></span>
<span><span class="co">## 3 M     M        79    412    0.119  0.192</span></span>
<span><span class="co">## 4 L     L       111    208    0.0600 0.534</span></span>
<span></span>
<span><span class="co"># Three different estimates:</span></span>
<span><span class="va">one_versus_all</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">summarize</span><span class="op">(</span></span>
<span>    macro <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">sens</span><span class="op">)</span>, </span>
<span>    macro_wts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/weighted.mean.html">weighted.mean</a></span><span class="op">(</span><span class="va">sens</span>, <span class="va">class_wts</span><span class="op">)</span>,</span>
<span>    micro <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">totals</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   macro macro_wts micro</span></span>
<span><span class="co">##   &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">## 1 0.560     0.709 0.709</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Thankfully, there is no need to manually implement these averaging methods. Instead, <span class="pkg">yardstick</span> functions can automatically apply these methods via the <code>estimator</code> argument:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sensitivity</span><span class="op">(</span><span class="va">hpc_cv</span>, <span class="va">obs</span>, <span class="va">pred</span>, estimator <span class="op">=</span> <span class="st">"macro"</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric     .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;       &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 sensitivity macro          0.560</span></span>
<span><span class="fu">sensitivity</span><span class="op">(</span><span class="va">hpc_cv</span>, <span class="va">obs</span>, <span class="va">pred</span>, estimator <span class="op">=</span> <span class="st">"macro_weighted"</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric     .estimator     .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;       &lt;chr&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">## 1 sensitivity macro_weighted     0.709</span></span>
<span><span class="fu">sensitivity</span><span class="op">(</span><span class="va">hpc_cv</span>, <span class="va">obs</span>, <span class="va">pred</span>, estimator <span class="op">=</span> <span class="st">"micro"</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric     .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;       &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 sensitivity micro          0.709</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When dealing with probability estimates, there are some metrics with multiclass analogs. For example, <span class="citation" data-cites="HandTill">Hand and Till (<a href="references.html#ref-HandTill" role="doc-biblioref">2001</a>)</span> determined a multiclass technique for ROC curves. In this case, <em>all</em> of the class probability columns must be given to the function:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">roc_auc</span><span class="op">(</span><span class="va">hpc_cv</span>, <span class="va">obs</span>, <span class="va">VF</span>, <span class="cn">F</span>, <span class="va">M</span>, <span class="va">L</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 roc_auc hand_till      0.829</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Macro-weighted averaging is also available as an option for applying this metric to a multiclass outcome:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">roc_auc</span><span class="op">(</span><span class="va">hpc_cv</span>, <span class="va">obs</span>, <span class="va">VF</span>, <span class="cn">F</span>, <span class="va">M</span>, <span class="va">L</span>, estimator <span class="op">=</span> <span class="st">"macro_weighted"</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 3</span></span>
<span><span class="co">##   .metric .estimator     .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">## 1 roc_auc macro_weighted     0.868</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, all of these performance metrics can be computed using <span class="pkg">dplyr</span> groupings. Recall that these data have a column for the resampling groups. We haven’t yet discussed resampling in detail, but notice how we can pass a grouped data frame to the metric function to compute the metrics for each group:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hpc_cv</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">Resample</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">accuracy</span><span class="op">(</span><span class="va">obs</span>, <span class="va">pred</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 10 × 4</span></span>
<span><span class="co">##   Resample .metric  .estimator .estimate</span></span>
<span><span class="co">##   &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;          &lt;dbl&gt;</span></span>
<span><span class="co">## 1 Fold01   accuracy multiclass     0.726</span></span>
<span><span class="co">## 2 Fold02   accuracy multiclass     0.712</span></span>
<span><span class="co">## 3 Fold03   accuracy multiclass     0.758</span></span>
<span><span class="co">## 4 Fold04   accuracy multiclass     0.712</span></span>
<span><span class="co">## 5 Fold05   accuracy multiclass     0.712</span></span>
<span><span class="co">## 6 Fold06   accuracy multiclass     0.697</span></span>
<span><span class="co">## # ℹ 4 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The groupings also translate to the <code>autoplot()</code> methods, with results shown in <a href="#fig-grouped-roc-curves">Figure&nbsp;<span>9.4</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Four 1-vs-all ROC curves for each fold</span></span>
<span><span class="va">hpc_cv</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">Resample</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">roc_curve</span><span class="op">(</span><span class="va">obs</span>, <span class="va">VF</span>, <span class="cn">F</span>, <span class="va">M</span>, <span class="va">L</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">autoplot</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-grouped-roc-curves" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="figures/fig-grouped-roc-curves-1.png" class="img-fluid figure-img" alt="Resampled ROC curves for each of the four outcome classes. There are four panels for classes VF, F, M, and L. Each panel contains ten ROC curves for each of the resampled data sets." width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;9.4: Resampled ROC curves for each of the four outcome classes</figcaption></figure>
</div>
</div>
</div>
<p>This visualization shows us that the different groups all perform about the same, but that the <code>VF</code> class is predicted better than the <code>F</code> or <code>M</code> classes, since the <code>VF</code> ROC curves are more in the top-left corner. This example uses resamples as the groups, but any grouping in your data can be used. This <code>autoplot()</code> method can be a quick visualization method for model effectiveness across outcome classes and/or groups.</p>
</section><section id="sec-performance-summary" class="level2" data-number="9.5"><h2 data-number="9.5" class="anchored" data-anchor-id="sec-performance-summary">
<span class="header-section-number">9.5</span> Chapter Summary</h2>
<p>Different metrics measure different aspects of a model fit, e.g., RMSE measures accuracy while the <span class="math inline">\(R^2\)</span> measures correlation. Measuring model performance is important even when a given model will not be used primarily for prediction; predictive power is also important for inferential or descriptive models. Functions from the <span class="pkg">yardstick</span> package measure the effectiveness of a model using data. The primary tidymodels interface uses tidyverse principles and data frames (as opposed to having vector arguments). Different metrics are appropriate for regression and classification metrics and, within these, there are sometimes different ways to estimate the statistics, such as for multiclass outcomes.</p>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-CraigSchapiro" class="csl-entry" role="listitem">
Craig–Schapiro, R, M Kuhn, C Xiong, E Pickering, J Liu, T Misko, R Perrin, et al. 2011. <span>“Multiplexed Immunoassay Panel Identifies Novel <span>CSF</span> Biomarkers for <span class="nocase">Alzheimer’s</span> Disease Diagnosis and Prognosis.”</span> <em>PLoS ONE</em> 6 (4): e18850.
</div>
<div id="ref-HandTill" class="csl-entry" role="listitem">
Hand, D, and R Till. 2001. <span>“A Simple Generalisation of the Area Under the <span>ROC</span> Curve for Multiple Class Classification Problems.”</span> <em>Machine Learning</em> 45 (August): 171–86.
</div>
<div id="ref-HosmerLemeshow" class="csl-entry" role="listitem">
Hosmer, D, and Sy Lemeshow. 2000. <em>Applied Logistic Regression</em>. New York: John Wiley; Sons.
</div>
<div id="ref-Kim:2009p4370" class="csl-entry" role="listitem">
Jungsu, K, D Basak, and D Holtzman. 2009. <span>“The Role of Apolipoprotein <span>E</span> in <span class="nocase">Alzheimer’s</span> Disease.”</span> <em>Neuron</em> 63 (3): 287–303.
</div>
<div id="ref-OpitzBurst" class="csl-entry" role="listitem">
Opitz, J, and S Burst. 2019. <span>“Macro F1 and Macro F1.”</span> <a href="https://arxiv.org/abs/1911.03347">https://arxiv.org/abs/1911.03347</a>.
</div>
<div id="ref-wu2017unified" class="csl-entry" role="listitem">
Wu, X, and Z Zhou. 2017. <span>“A Unified View of Multi-Label Performance Measures.”</span> In <em>International Conference on Machine Learning</em>, 3780–88.
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./08-feature-engineering.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Feature Engineering with recipes</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./10-resampling.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Resampling for Evaluating Performance</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Modelado Ordenado con R fue escrito por Max Kuhn, y Julia Silge.</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Este libro fue creado usando <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>


</body></html>