<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Max Kuhn, y Julia Silge">
<title>Modelado Ordenado con R - 13&nbsp; Grid Search</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./14-iterative-search.html" rel="next">
<link href="./12-tuning-parameters.html" rel="prev">
<link href="./images/cover.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><link rel="stylesheet" href="style.css">
<link rel="stylesheet" href="TMwR.css">
</head>
<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./10-resampling.html">TOOLS FOR CREATING EFFECTIVE MODELS</a></li><li class="breadcrumb-item"><a href="./13-grid-search.html"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Grid Search</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Modelado Ordenado con R</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/davidrsch/TMwRes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hello World</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">INTRODUCTION</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-software-modeling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Software for modeling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">A Tidyverse Primer</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-base-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">A Review of R Modeling Fundamentals</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">MODELING BASICS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-ames.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">The Ames Housing Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-data-spending.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Spending our Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-fitting-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Fitting Models with parsnip</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-the-model-workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">A Model Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-feature-engineering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Feature Engineering with recipes</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-judging-model-effectiveness.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Judging Model Effectiveness</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">TOOLS FOR CREATING EFFECTIVE MODELS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-resampling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Resampling for Evaluating Performance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-comparing-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Comparing Models with Resampling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12-tuning-parameters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Model Tuning and the Dangers of Overfitting</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13-grid-search.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Grid Search</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14-iterative-search.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Iterative Search</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./15-workflow-sets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Screening Many Models</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">BEYOND THE BASICS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./16-dimensionality-reduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Dimensionality Reduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./17-encoding-categorical-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Encoding Categorical Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./18-explaining-models-and-predictions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Explaining Models and Predictions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./19-when-should-you-trust-predictions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">When Should You Trust Your Predictions?</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./20-ensemble-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Ensembles of Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21-inferential-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Inferential Analysis</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pre-proc-table.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Recommended Preprocessing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#sec-grids" id="toc-sec-grids" class="nav-link active" data-scroll-target="#sec-grids"><span class="header-section-number">13.1</span> Regular and Nonregular Grids</a>
  <ul class="collapse">
<li><a href="#regular-grids" id="toc-regular-grids" class="nav-link" data-scroll-target="#regular-grids">Regular grids</a></li>
  <li><a href="#irregular-grids" id="toc-irregular-grids" class="nav-link" data-scroll-target="#irregular-grids">Irregular grids</a></li>
  </ul>
</li>
  <li><a href="#sec-evaluating-grid" id="toc-sec-evaluating-grid" class="nav-link" data-scroll-target="#sec-evaluating-grid"><span class="header-section-number">13.2</span> Evaluating the Grid</a></li>
  <li><a href="#finalizing-the-model" id="toc-finalizing-the-model" class="nav-link" data-scroll-target="#finalizing-the-model"><span class="header-section-number">13.3</span> Finalizing the Model</a></li>
  <li><a href="#sec-tuning-usemodels" id="toc-sec-tuning-usemodels" class="nav-link" data-scroll-target="#sec-tuning-usemodels"><span class="header-section-number">13.4</span> Tools for Creating Tuning Specifications</a></li>
  <li>
<a href="#sec-efficient-grids" id="toc-sec-efficient-grids" class="nav-link" data-scroll-target="#sec-efficient-grids"><span class="header-section-number">13.5</span> Tools for Efficient Grid Search</a>
  <ul class="collapse">
<li><a href="#sec-submodel-trick" id="toc-sec-submodel-trick" class="nav-link" data-scroll-target="#sec-submodel-trick"><span class="header-section-number">13.5.1</span> Submodel optimization</a></li>
  <li><a href="#parallel-processing" id="toc-parallel-processing" class="nav-link" data-scroll-target="#parallel-processing"><span class="header-section-number">13.5.2</span> Parallel processing</a></li>
  <li><a href="#benchmarking-boosted-trees" id="toc-benchmarking-boosted-trees" class="nav-link" data-scroll-target="#benchmarking-boosted-trees"><span class="header-section-number">13.5.3</span> Benchmarking boosted trees</a></li>
  <li><a href="#access-to-global-variables" id="toc-access-to-global-variables" class="nav-link" data-scroll-target="#access-to-global-variables"><span class="header-section-number">13.5.4</span> Access to global variables</a></li>
  <li><a href="#sec-racing" id="toc-sec-racing" class="nav-link" data-scroll-target="#sec-racing"><span class="header-section-number">13.5.5</span> Racing methods</a></li>
  </ul>
</li>
  <li><a href="#sec-grid-summary" id="toc-sec-grid-summary" class="nav-link" data-scroll-target="#sec-grid-summary"><span class="header-section-number">13.6</span> Chapter Summary</a></li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/davidrsch/TMwRes/edit/main/13-grid-search.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/davidrsch/TMwRes/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-grid-search" class="quarto-section-identifier"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Grid Search</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p>In <a href="#sec-tuning">Chapter&nbsp;<span class="quarto-unresolved-ref ref-noprefix">sec-tuning</span></a> we demonstrated how users can mark or tag arguments in preprocessing recipes and/or model specifications for optimization using the <code><a href="https://hardhat.tidymodels.org/reference/tune.html">tune()</a></code> function. Once we know what to optimize, it’s time to address the question of how to optimize the parameters. This chapter describes <em>grid search</em> methods that specify the possible values of the parameters <em>a priori</em>. <a href="#sec-iterative-search">Chapter&nbsp;<span class="quarto-unresolved-ref ref-noprefix">sec-iterative-search</span></a> will continue the discussion by describing iterative search methods.)</p>
<p>Let’s start by looking at two main approaches for assembling a grid.</p>
<section id="sec-grids" class="level2" data-number="13.1"><h2 data-number="13.1" class="anchored" data-anchor-id="sec-grids">
<span class="header-section-number">13.1</span> Regular and Nonregular Grids</h2>
<p>There are two main types of grids. A regular grid combines each parameter (with its corresponding set of possible values) factorially, i.e., by using all combinations of the sets. Alternatively, a nonregular grid is one where the parameter combinations are not formed from a small set of points.</p>
<p>Before we look at each type in more detail, let’s consider an example model: the multilayer perceptron model (a.k.a. single layer artificial neural network). The parameters marked for tuning are:</p>
<ul>
<li><p>the number of hidden units</p></li>
<li><p>the number of fitting epochs/iterations in model training</p></li>
<li><p>the amount of weight decay penalization</p></li>
</ul>
<div class="rmdnote">
<p>Historically, the number of epochs was determined by early stopping; a separate validation set determined the length of training based on the error rate, since re-predicting the training set led to overfitting. In our case, the use of a weight decay penalty should prohibit overfitting, and there is little harm in tuning the penalty and the number of epochs.</p>
</div>
<p>Using <span class="pkg">parsnip</span>, the specification for a classification model fit using the <span class="pkg">nnet</span> package is:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-mlp_e7df18bce5bd2d42e80fd70823df9513">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidymodels.tidymodels.org">tidymodels</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://tidymodels.tidymodels.org/reference/tidymodels_prefer.html">tidymodels_prefer</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mlp_spec</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">mlp</span><span class="op">(</span>hidden_units <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, penalty <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, epochs <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"nnet"</span>, trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"classification"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The argument <code>trace = 0</code> prevents extra logging of the training process. As shown in <a href="#sec-tuning-params-tidymodels"><span class="quarto-unresolved-ref">sec-tuning-params-tidymodels</span></a>, the <code><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_parameter_set_dials()</a></code> function can extract the set of arguments with unknown values and sets their <span class="pkg">dials</span> objects:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-mlp-param_7a3775f11506a5b3cd8c5caf0084364d">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mlp_param</span> <span class="op">&lt;-</span> <span class="fu">extract_parameter_set_dials</span><span class="op">(</span><span class="va">mlp_spec</span><span class="op">)</span></span>
<span><span class="va">mlp_param</span> <span class="op">%&gt;%</span> <span class="fu">extract_parameter_dials</span><span class="op">(</span><span class="st">"hidden_units"</span><span class="op">)</span></span>
<span><span class="co">## # Hidden Units (quantitative)</span></span>
<span><span class="co">## Range: [1, 10]</span></span>
<span><span class="va">mlp_param</span> <span class="op">%&gt;%</span> <span class="fu">extract_parameter_dials</span><span class="op">(</span><span class="st">"penalty"</span><span class="op">)</span></span>
<span><span class="co">## Amount of Regularization (quantitative)</span></span>
<span><span class="co">## Transformer: log-10 [1e-100, Inf]</span></span>
<span><span class="co">## Range (transformed scale): [-10, 0]</span></span>
<span><span class="va">mlp_param</span> <span class="op">%&gt;%</span> <span class="fu">extract_parameter_dials</span><span class="op">(</span><span class="st">"epochs"</span><span class="op">)</span></span>
<span><span class="co">## # Epochs (quantitative)</span></span>
<span><span class="co">## Range: [10, 1000]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This output indicates that the parameter objects are complete and prints their default ranges. These values will be used to demonstrate how to create different types of parameter grids.</p>
<section id="regular-grids" class="level3 unnumbered"><h3 class="unnumbered anchored" data-anchor-id="regular-grids">Regular grids</h3>
<p>Regular grids are combinations of separate sets of parameter values. First, the user creates a distinct set of values for each parameter. The number of possible values need not be the same for each parameter. The <span class="pkg">tidyr</span> function <code>crossing()</code> is one way to create a regular grid:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-crossing_b72b63f3f6e442bf8b46cced93d413f1">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">crossing</span><span class="op">(</span></span>
<span>  hidden_units <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>  penalty <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.0</span>, <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  epochs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">200</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 12 × 3</span></span>
<span><span class="co">##   hidden_units penalty epochs</span></span>
<span><span class="co">##          &lt;int&gt;   &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">## 1            1     0      100</span></span>
<span><span class="co">## 2            1     0      200</span></span>
<span><span class="co">## 3            1     0.1    100</span></span>
<span><span class="co">## 4            1     0.1    200</span></span>
<span><span class="co">## 5            2     0      100</span></span>
<span><span class="co">## 6            2     0      200</span></span>
<span><span class="co">## # ℹ 6 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The parameter object knows the ranges of the parameters. The <span class="pkg">dials</span> package contains a set of <code>grid_*()</code> functions that take the parameter object as input to produce different types of grids. For example:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-reg_77cdd3e13c920ac5bd06ec68aa98cad2">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">grid_regular</span><span class="op">(</span><span class="va">mlp_param</span>, levels <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 8 × 3</span></span>
<span><span class="co">##   hidden_units      penalty epochs</span></span>
<span><span class="co">##          &lt;int&gt;        &lt;dbl&gt;  &lt;int&gt;</span></span>
<span><span class="co">## 1            1 0.0000000001     10</span></span>
<span><span class="co">## 2           10 0.0000000001     10</span></span>
<span><span class="co">## 3            1 1                10</span></span>
<span><span class="co">## 4           10 1                10</span></span>
<span><span class="co">## 5            1 0.0000000001   1000</span></span>
<span><span class="co">## 6           10 0.0000000001   1000</span></span>
<span><span class="co">## # ℹ 2 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>levels</code> argument is the number of levels per parameter to create. It can also take a named vector of values:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-reg-lvls_5efe11cd8635c404ea42c3f6471b7da8">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mlp_param</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">grid_regular</span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>hidden_units <span class="op">=</span> <span class="fl">3</span>, penalty <span class="op">=</span> <span class="fl">2</span>, epochs <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 12 × 3</span></span>
<span><span class="co">##   hidden_units      penalty epochs</span></span>
<span><span class="co">##          &lt;int&gt;        &lt;dbl&gt;  &lt;int&gt;</span></span>
<span><span class="co">## 1            1 0.0000000001     10</span></span>
<span><span class="co">## 2            5 0.0000000001     10</span></span>
<span><span class="co">## 3           10 0.0000000001     10</span></span>
<span><span class="co">## 4            1 1                10</span></span>
<span><span class="co">## 5            5 1                10</span></span>
<span><span class="co">## 6           10 1                10</span></span>
<span><span class="co">## # ℹ 6 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are techniques for creating regular grids that do not use all possible values of each parameter set. These <em>fractional factorial designs</em> <span class="citation" data-cites="BHH">(<a href="#ref-BHH" role="doc-biblioref">Box, Hunter, and Hunter 2005</a>)</span> could also be used. To learn more, consult the CRAN Task View for experimental design.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<div class="rmdwarning">
<p>Regular grids can be computationally expensive to use, especially when there are a medium-to-large number of tuning parameters. This is true for many models but not all. As discussed in <a href="#sec-efficient-grids"><span class="quarto-unresolved-ref">sec-efficient-grids</span></a> below, there are many models whose tuning time <em>decreases</em> with a regular grid!</p>
</div>
<p>One advantage to using a regular grid is that the relationships and patterns between the tuning parameters and the model metrics are easily understood. The factorial nature of these designs allows for examination of each parameter separately with little confounding between parameters.</p>
</section><section id="irregular-grids" class="level3 unnumbered"><h3 class="unnumbered anchored" data-anchor-id="irregular-grids">Irregular grids</h3>
<p>There are several options for creating non-regular grids. The first is to use random sampling across the range of parameters. The <code>grid_random()</code> function generates independent uniform random numbers across the parameter ranges. If the parameter object has an associated transformation (such as we have for <code>penalty</code>), the random numbers are generated on the transformed scale. Let’s create a random grid for the parameters from our example neural network:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-rand_be492c3a1118368e97bad4d738d77a02">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1301</span><span class="op">)</span></span>
<span><span class="va">mlp_param</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">grid_random</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># 'size' is the number of combinations</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##   hidden_units      penalty           epochs   </span></span>
<span><span class="co">##  Min.   : 1.00   Min.   :0.0000   Min.   : 10  </span></span>
<span><span class="co">##  1st Qu.: 3.00   1st Qu.:0.0000   1st Qu.:266  </span></span>
<span><span class="co">##  Median : 5.00   Median :0.0000   Median :497  </span></span>
<span><span class="co">##  Mean   : 5.38   Mean   :0.0437   Mean   :510  </span></span>
<span><span class="co">##  3rd Qu.: 8.00   3rd Qu.:0.0027   3rd Qu.:761  </span></span>
<span><span class="co">##  Max.   :10.00   Max.   :0.9814   Max.   :999</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For <code>penalty</code>, the random numbers are uniform on the log (base-10) scale but the values in the grid are in the natural units.</p>
<p>The issue with random grids is that, with small-to-medium grids, random values can result in overlapping parameter combinations. Also, the random grid needs to cover the whole parameter space, but the likelihood of good coverage increases with the number of grid values. Even for a sample of 15 candidate points, <a href="#fig-random-grid">Figure&nbsp;<span class="quarto-unresolved-ref">fig-random-grid</span></a> shows some overlap between points for our example multilayer perceptron.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-random-matrix_68a9aa5a7902769048bbc10da8251277">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggforce.data-imaginist.com">ggforce</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1302</span><span class="op">)</span></span>
<span><span class="va">mlp_param</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="co"># The 'original = FALSE' option keeps penalty in log10 units</span></span>
<span>  <span class="fu">grid_random</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span>, original <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.panel_x</span>, y <span class="op">=</span> <span class="va">.panel_y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_blank.html">geom_blank</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggforce.data-imaginist.com/reference/facet_matrix.html">facet_matrix</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">hidden_units</span>, <span class="va">penalty</span>, <span class="va">epochs</span><span class="op">)</span>, layer.diag <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Random design with 20 candidates"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-random-grid_f26241f813e232af324364b3131130cc">
<div class="cell-output-display">
<div id="fig-random-grid" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-random-grid-1.png" class="img-fluid figure-img" alt="A scatter plot matrix for three tuning parameters with 20 points generated at random. There are significant gaps in the parameter space." width="576"></p>
<figcaption class="figure-caption">Figure&nbsp;13.1: Three tuning parameters with 15 points generated at random</figcaption></figure>
</div>
</div>
</div>
<p>A much better approach is to use a set of experimental designs called <em>space-filling designs</em>. While different design methods have slightly different goals, they generally find a configuration of points that cover the parameter space with the smallest chance of overlapping or redundant values. Examples of such designs are Latin hypercubes <span class="citation" data-cites="lhd">(<a href="#ref-lhd" role="doc-biblioref">McKay, Beckman, and Conover 1979</a>)</span>, maximum entropy designs <span class="citation" data-cites="maxent">(<a href="#ref-maxent" role="doc-biblioref">Shewry and Wynn 1987</a>)</span>, maximum projection designs <span class="citation" data-cites="maxproj">(<a href="#ref-maxproj" role="doc-biblioref">Joseph, Gul, and Ba 2015</a>)</span>, and others. See <span class="citation" data-cites="santner2003design">Santner et al. (<a href="#ref-santner2003design" role="doc-biblioref">2003</a>)</span> for an overview.</p>
<p>The <span class="pkg">dials</span> package contains functions for Latin hypercube and maximum entropy designs. As with <code>grid_random()</code>, the primary inputs are the number of parameter combinations and a parameter object. Let’s compare a random design with a Latin hypercube design for 20 candidate parameter values in <a href="#fig-space-filling-design">Figure&nbsp;<span class="quarto-unresolved-ref">fig-space-filling-design</span></a>.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-sfd-compare_95693d6085883ecedd4b6df970dcfdca">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1303</span><span class="op">)</span></span>
<span><span class="va">mlp_param</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">grid_latin_hypercube</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span>, original <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.panel_x</span>, y <span class="op">=</span> <span class="va">.panel_y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_blank.html">geom_blank</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggforce.data-imaginist.com/reference/facet_matrix.html">facet_matrix</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">hidden_units</span>, <span class="va">penalty</span>, <span class="va">epochs</span><span class="op">)</span>, layer.diag <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Latin Hypercube design with 20 candidates"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-space-filling-design_9084b5ddbb7b5965353ddb70663889b8">
<div class="cell-output-display">
<div id="fig-space-filling-design" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-space-filling-design-1.png" class="img-fluid figure-img" alt="A scatter plot matrix for three tuning parameters with 15 points generated using a space-filling design. There are fewer gaps in the parameter space when compared to the random grid." width="576"></p>
<figcaption class="figure-caption">Figure&nbsp;13.2: Three tuning parameters with 20 points generated using a space-filling design</figcaption></figure>
</div>
</div>
</div>
<p>While not perfect, this Latin hypercube design spaces the points farther away from one another and allows a better exploration of the hyperparameter space.</p>
<p>Space-filling designs can be very effective at representing the parameter space. The default design used by the <span class="pkg">tune</span> package is the maximum entropy design. These tend to produce grids that cover the candidate space well and drastically increase the chances of finding good results.</p>
</section></section><section id="sec-evaluating-grid" class="level2" data-number="13.2"><h2 data-number="13.2" class="anchored" data-anchor-id="sec-evaluating-grid">
<span class="header-section-number">13.2</span> Evaluating the Grid</h2>
<p>To choose the best tuning parameter combination, each candidate set is assessed using data that were not used to train that model. Resampling methods or a single validation set work well for this purpose. The process (and syntax) closely resembles the approach in <a href="#sec-resampling-performance"><span class="quarto-unresolved-ref">sec-resampling-performance</span></a> that used the <code><a href="https://tune.tidymodels.org/reference/fit_resamples.html">fit_resamples()</a></code> function from the <span class="pkg">tune</span> package.</p>
<p>After resampling, the user selects the most appropriate candidate parameter set. It might make sense to choose the empirically best parameter combination or bias the choice towards other aspects of the model fit, such as simplicity.</p>
<p>We use a classification data set to demonstrate model tuning in this and the next chapter. The data come from <span class="citation" data-cites="Hill">Hill et al. (<a href="#ref-Hill" role="doc-biblioref">2007</a>)</span>, who developed an automated microscopy laboratory tool for cancer research. The data consists of 56 imaging measurements on 2019 human breast cancer cells. These predictors represent shape and intensity characteristics of different parts of the cells (e.g., the nucleus, the cell boundary, etc.). There is a high degree of correlation between the predictors. For example, there are several different predictors that measure the size and shape of the nucleus and cell boundary. Also, individually, many predictors have skewed distributions.</p>
<p>Each cell belongs to one of two classes. Since this is part of an automated lab test, the focus was on prediction capability rather than inference.</p>
<p>The data are included in the <span class="pkg">modeldata</span> package. Let’s remove one column not needed for analysis (<code>case</code>):</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells_71bc187eeca78926b736cc7428519a98">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidymodels.tidymodels.org">tidymodels</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="va">cells</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">case</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Given the dimensions of the data, we can compute performance metrics using 10-fold cross-validation:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-folds_434642c1474f21ef87f1c56b9acf3e16">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1304</span><span class="op">)</span></span>
<span><span class="va">cell_folds</span> <span class="op">&lt;-</span> <span class="fu">vfold_cv</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Because of the high degree of correlation between predictors, it makes sense to use PCA feature extraction to decorrelate the predictors. The following recipe contains steps to transform the predictors to increase symmetry, normalize them to be on the same scale, then conduct feature extraction. The number of PCA components to retain is also tuned, along with the model parameters.</p>
<div class="rmdwarning">
<p>While the resulting PCA components are technically on the same scale, the lower-rank components tend to have a wider range than the higher-rank components. For this reason, we normalize again to coerce the predictors to have the same mean and variance.</p>
</div>
<p>Many of the predictors have skewed distributions. Since PCA is variance based, extreme values can have a detrimental effect on these calculations. To counter this, let’s add a recipe step estimating a Yeo-Johnson transformation for each predictor <span class="citation" data-cites="yeo2000new">(<a href="#ref-yeo2000new" role="doc-biblioref">Yeo and Johnson 2000</a>)</span>. While originally intended as a transformation of the outcome, it can also be used to estimate transformations that encourage more symmetric distributions. This step <code>step_YeoJohnson()</code> occurs in the recipe just prior to the initial normalization via <code>step_normalize()</code>. Then, let’s combine this feature engineering recipe with our neural network model specification <code>mlp_spec</code>.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-objects_a2e1845dd5f494ba49f64151ebc19116">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mlp_rec</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">recipe</span><span class="op">(</span><span class="va">class</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">cells</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">step_YeoJohnson</span><span class="op">(</span><span class="fu">all_numeric_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">step_normalize</span><span class="op">(</span><span class="fu">all_numeric_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">step_pca</span><span class="op">(</span><span class="fu">all_numeric_predictors</span><span class="op">(</span><span class="op">)</span>, num_comp <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">step_normalize</span><span class="op">(</span><span class="fu">all_numeric_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mlp_wflow</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">workflow</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">add_model</span><span class="op">(</span><span class="va">mlp_spec</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">add_recipe</span><span class="op">(</span><span class="va">mlp_rec</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s create a parameter object <code>mlp_param</code> to adjust a few of the default ranges. We can change the number of epochs to have a smaller range (50 to 200 epochs). Also, the default range for <code>num_comp()</code> defaults to a very narrow range (one to four components); we can increase the range to 40 components and set the minimum value to zero:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-workflow_65b9ccfb947e4538aa64ca5bf151ebcc">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mlp_param</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">mlp_wflow</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">extract_parameter_set_dials</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">update</span><span class="op">(</span></span>
<span>    epochs <span class="op">=</span> <span class="fu">epochs</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    num_comp <span class="op">=</span> <span class="fu">num_comp</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">40</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="rmdnote">
<p>In <code>step_pca()</code>, using zero PCA components is a shortcut to skip the feature extraction. In this way, the original predictors can be directly compared to the results that include PCA components.</p>
</div>
<p>The <code><a href="https://tune.tidymodels.org/reference/tune_grid.html">tune_grid()</a></code> function is the primary function for conducting grid search. Its functionality is very similar to <code><a href="https://tune.tidymodels.org/reference/fit_resamples.html">fit_resamples()</a></code> from <a href="#sec-resampling-performance"><span class="quarto-unresolved-ref">sec-resampling-performance</span></a>, although it has additional arguments related to the grid:</p>
<ul>
<li><p><code>grid</code>: An integer or data frame. When an integer is used, the function creates a space-filling design with <code>grid</code> number of candidate parameter combinations. If specific parameter combinations exist, the <code>grid</code> parameter is used to pass them to the function.</p></li>
<li><p><code>param_info</code>: An optional argument for defining the parameter ranges. The argument is most useful when <code>grid</code> is an integer.</p></li>
</ul>
<p>Otherwise, the interface to <code><a href="https://tune.tidymodels.org/reference/tune_grid.html">tune_grid()</a></code> is the same as <code><a href="https://tune.tidymodels.org/reference/fit_resamples.html">fit_resamples()</a></code>. The first argument is either a model specification or workflow. When a model is given, the second argument can be either a recipe or formula. The other required argument is an <span class="pkg">rsample</span> resampling object (such as <code>cell_folds</code>). The following call also passes a metric set so that the area under the ROC curve is measured during resampling.</p>
<p>To start, let’s evaluate a regular grid with three levels across the resamples:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-regular_785d7594690fa4308051482d7aa587be">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">roc_res</span> <span class="op">&lt;-</span> <span class="fu">metric_set</span><span class="op">(</span><span class="va">roc_auc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1305</span><span class="op">)</span></span>
<span><span class="va">mlp_reg_tune</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">mlp_wflow</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">tune_grid</span><span class="op">(</span></span>
<span>    <span class="va">cell_folds</span>,</span>
<span>    grid <span class="op">=</span> <span class="va">mlp_param</span> <span class="op">%&gt;%</span> <span class="fu">grid_regular</span><span class="op">(</span>levels <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>,</span>
<span>    metrics <span class="op">=</span> <span class="va">roc_res</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">mlp_reg_tune</span></span>
<span><span class="co">## # Tuning results</span></span>
<span><span class="co">## # 10-fold cross-validation </span></span>
<span><span class="co">## # A tibble: 10 × 4</span></span>
<span><span class="co">##   splits             id     .metrics          .notes          </span></span>
<span><span class="co">##   &lt;list&gt;             &lt;chr&gt;  &lt;list&gt;            &lt;list&gt;          </span></span>
<span><span class="co">## 1 &lt;split [1817/202]&gt; Fold01 &lt;tibble [81 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 2 &lt;split [1817/202]&gt; Fold02 &lt;tibble [81 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 3 &lt;split [1817/202]&gt; Fold03 &lt;tibble [81 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 4 &lt;split [1817/202]&gt; Fold04 &lt;tibble [81 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 5 &lt;split [1817/202]&gt; Fold05 &lt;tibble [81 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 6 &lt;split [1817/202]&gt; Fold06 &lt;tibble [81 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## # ℹ 4 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are high-level convenience functions we can use to understand the results. First, the <code><a href="https://ggplot2.tidyverse.org/reference/autoplot.html">autoplot()</a></code> method for regular grids shows the performance profiles across tuning parameters in <a href="#fig-regular-grid-plot">Figure&nbsp;<span class="quarto-unresolved-ref">fig-regular-grid-plot</span></a>.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-reg-plot_2a5fe7c12bed6c55f85ed316ab254be0">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html">autoplot</a></span><span class="op">(</span><span class="va">mlp_reg_tune</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_color_viridis_d</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-regular-grid-plot_6727c74589b5a0aa05b28b2c35c05c19">
<div class="cell-output-display">
<div id="fig-regular-grid-plot" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-regular-grid-plot-1.png" class="img-fluid figure-img" alt="A line plot of the regular grid results. The x axis shows the number of hidden units and the y axis is the resampled ROC AUC. There are separate lines for the amount of regularization. There are nine panels for three values for the number of PCA components and the number of epochs. On average, the amount of regularization is important where more is better. Also, on average, the increasing the number of hidden units decreases model effectiveness." width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;13.3: The regular grid results</figcaption></figure>
</div>
</div>
</div>
<p>For these data, the amount of penalization has the largest impact on the area under the ROC curve. The number of epochs doesn’t appear to have a pronounced effect on performance. The change in the number of hidden units appears to matter most when the amount of regularization is low (and harms performance). There are several parameter configurations that have roughly equivalent performance, as seen using the function <code><a href="https://tune.tidymodels.org/reference/show_best.html">show_best()</a></code>:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-reg-best_a9608818c6423a9a7bec8b47f0c55f77">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">show_best</span><span class="op">(</span><span class="va">mlp_reg_tune</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">.estimator</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 5 × 9</span></span>
<span><span class="co">##   hidden_units penalty epochs num_comp .metric  mean     n std_err .config          </span></span>
<span><span class="co">##          &lt;int&gt;   &lt;dbl&gt;  &lt;int&gt;    &lt;int&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;            </span></span>
<span><span class="co">## 1            5       1     50        0 roc_auc 0.897    10 0.00857 Preprocessor1_Mo…</span></span>
<span><span class="co">## 2           10       1    125        0 roc_auc 0.895    10 0.00898 Preprocessor1_Mo…</span></span>
<span><span class="co">## 3           10       1     50        0 roc_auc 0.894    10 0.00960 Preprocessor1_Mo…</span></span>
<span><span class="co">## 4            5       1    200        0 roc_auc 0.894    10 0.00784 Preprocessor1_Mo…</span></span>
<span><span class="co">## 5            5       1    125        0 roc_auc 0.892    10 0.00822 Preprocessor1_Mo…</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Based on these results, it would make sense to conduct another run of grid search with larger values of the weight decay penalty.</p>
<p>To use a space-filling design, either the <code>grid</code> argument can be given an integer or one of the <code>grid_*()</code> functions can produce a data frame. To evaluate the same range using a maximum entropy design with 20 candidate values:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-sdf_e6d794adfb102804ca856fcfdff2f4b9">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1306</span><span class="op">)</span></span>
<span><span class="va">mlp_sfd_tune</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">mlp_wflow</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">tune_grid</span><span class="op">(</span></span>
<span>    <span class="va">cell_folds</span>,</span>
<span>    grid <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    <span class="co"># Pass in the parameter object to use the appropriate range: </span></span>
<span>    param_info <span class="op">=</span> <span class="va">mlp_param</span>,</span>
<span>    metrics <span class="op">=</span> <span class="va">roc_res</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">mlp_sfd_tune</span></span>
<span><span class="co">## # Tuning results</span></span>
<span><span class="co">## # 10-fold cross-validation </span></span>
<span><span class="co">## # A tibble: 10 × 4</span></span>
<span><span class="co">##   splits             id     .metrics          .notes          </span></span>
<span><span class="co">##   &lt;list&gt;             &lt;chr&gt;  &lt;list&gt;            &lt;list&gt;          </span></span>
<span><span class="co">## 1 &lt;split [1817/202]&gt; Fold01 &lt;tibble [20 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 2 &lt;split [1817/202]&gt; Fold02 &lt;tibble [20 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 3 &lt;split [1817/202]&gt; Fold03 &lt;tibble [20 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 4 &lt;split [1817/202]&gt; Fold04 &lt;tibble [20 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 5 &lt;split [1817/202]&gt; Fold05 &lt;tibble [20 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## 6 &lt;split [1817/202]&gt; Fold06 &lt;tibble [20 × 8]&gt; &lt;tibble [0 × 3]&gt;</span></span>
<span><span class="co">## # ℹ 4 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://ggplot2.tidyverse.org/reference/autoplot.html">autoplot()</a></code> method will also work with these designs, although the format of the results will be different. <a href="#fig-sfd-plot">Figure&nbsp;<span class="quarto-unresolved-ref">fig-sfd-plot</span></a> was produced using <code>autoplot(mlp_sfd_tune)</code>.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-sfd-plot_d6899b0dd6eb6a124c578070468c47f5">
<div class="cell-output-display">
<div id="fig-sfd-plot" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-sfd-plot-1.png" class="img-fluid figure-img" alt="The `autoplot()` method results when used with a space-filling design. The trends show decreasing performance with the number of PCA components as well as the number of hidden units." width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;13.4: The <code><a href="https://ggplot2.tidyverse.org/reference/autoplot.html">autoplot()</a></code> method results when used with a space-filling design</figcaption></figure>
</div>
</div>
</div>
<p>This marginal effects plot (<a href="#fig-sfd-plot">Figure&nbsp;<span class="quarto-unresolved-ref">fig-sfd-plot</span></a>) shows the relationship of each parameter with the performance metric.</p>
<div class="rmdwarning">
<p>Take care when examining this plot; since a regular grid is not used, the values of the other tuning parameters can affect each panel.</p>
</div>
<p>The penalty parameter appears to result in better performance with smaller amounts of weight decay. This is the opposite of the results from the regular grid. Since each point in each panel is shared with the other three tuning parameters, the trends in one panel can be affected by the others. Using a regular grid, each point in each panel is equally averaged over the other parameters. For this reason, the effect of each parameter is better isolated with regular grids.</p>
<p>As with the regular grid, <code><a href="https://tune.tidymodels.org/reference/show_best.html">show_best()</a></code> can report on the numerically best results:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-sdf-best_20a342ae3c723ba946e5ef3ce5270908">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">show_best</span><span class="op">(</span><span class="va">mlp_sfd_tune</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">.estimator</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 5 × 9</span></span>
<span><span class="co">##   hidden_units       penalty epochs num_comp .metric  mean     n std_err .config    </span></span>
<span><span class="co">##          &lt;int&gt;         &lt;dbl&gt;  &lt;int&gt;    &lt;int&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;      </span></span>
<span><span class="co">## 1            8 0.594             97       22 roc_auc 0.880    10 0.00998 Preprocess…</span></span>
<span><span class="co">## 2            3 0.00000000649    135        8 roc_auc 0.878    10 0.00953 Preprocess…</span></span>
<span><span class="co">## 3            9 0.141            177       11 roc_auc 0.873    10 0.0104  Preprocess…</span></span>
<span><span class="co">## 4            8 0.0000000103      74        9 roc_auc 0.869    10 0.00761 Preprocess…</span></span>
<span><span class="co">## 5            6 0.00581          129       15 roc_auc 0.865    10 0.00658 Preprocess…</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Generally, it is a good idea to evaluate the models over multiple metrics so that different aspects of the model fit are taken into account. Also, it often makes sense to choose a slightly suboptimal parameter combination that is associated with a simpler model. For this model, simplicity corresponds to larger penalty values and/or fewer hidden units.</p>
<p>As with the results from <code><a href="https://tune.tidymodels.org/reference/fit_resamples.html">fit_resamples()</a></code>, there is usually no value in retaining the intermediary model fits across the resamples and tuning parameters. However, as before, the <code>extract</code> option to <code><a href="https://tune.tidymodels.org/reference/control_grid.html">control_grid()</a></code> allows the retention of the fitted models and/or recipes. Also, setting the <code>save_pred</code> option to <code>TRUE</code> retains the assessment set predictions and these can be accessed using <code><a href="https://tune.tidymodels.org/reference/collect_predictions.html">collect_predictions()</a></code>.</p>
</section><section id="finalizing-the-model" class="level2" data-number="13.3"><h2 data-number="13.3" class="anchored" data-anchor-id="finalizing-the-model">
<span class="header-section-number">13.3</span> Finalizing the Model</h2>
<p>If one of the sets of possible model parameters found via <code><a href="https://tune.tidymodels.org/reference/show_best.html">show_best()</a></code> were an attractive final option for these data, we might wish to evaluate how well it does on the test set. However, the results of <code><a href="https://tune.tidymodels.org/reference/tune_grid.html">tune_grid()</a></code> only provide the substrate to choose appropriate tuning parameters. The function <em>does not fit</em> a final model.</p>
<p>To fit a final model, a final set of parameter values must be determined. There are two methods to do so:</p>
<ul>
<li>manually pick values that appear appropriate or</li>
<li>use a <code>select_*()</code> function.</li>
</ul>
<p>For example, <code><a href="https://tune.tidymodels.org/reference/show_best.html">select_best()</a></code> will choose the parameters with the numerically best results. Let’s go back to our regular grid results and see which one is best:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-select_6bc7a6fcfdede8b8d8e97c953f10bd90">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">select_best</span><span class="op">(</span><span class="va">mlp_reg_tune</span>, metric <span class="op">=</span> <span class="st">"roc_auc"</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 5</span></span>
<span><span class="co">##   hidden_units penalty epochs num_comp .config              </span></span>
<span><span class="co">##          &lt;int&gt;   &lt;dbl&gt;  &lt;int&gt;    &lt;int&gt; &lt;chr&gt;                </span></span>
<span><span class="co">## 1            5       1     50        0 Preprocessor1_Model08</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Looking back at <a href="#fig-regular-grid-plot">Figure&nbsp;<span class="quarto-unresolved-ref">fig-regular-grid-plot</span></a>, we can see that a model with a single hidden unit trained for 125 epochs on the original predictors with a large amount of penalization has performance competitive with this option, and is simpler. This is basically penalized logistic regression! To manually specify these parameters, we can create a tibble with these values and then use a <em>finalization</em> function to splice the values back into the workflow:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-finalize-manual_c2a9e711a933410cf1ee6dad8a460ddb">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">logistic_param</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">tibble</span><span class="op">(</span></span>
<span>    num_comp <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    epochs <span class="op">=</span> <span class="fl">125</span>,</span>
<span>    hidden_units <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    penalty <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">final_mlp_wflow</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">mlp_wflow</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">finalize_workflow</span><span class="op">(</span><span class="va">logistic_param</span><span class="op">)</span></span>
<span><span class="va">final_mlp_wflow</span></span>
<span><span class="co">## ══ Workflow ═════════════════════════════════════════════════════════════════════════</span></span>
<span><span class="co">## Preprocessor: Recipe</span></span>
<span><span class="co">## Model: mlp()</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ── Preprocessor ─────────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">## 4 Recipe Steps</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## • step_YeoJohnson()</span></span>
<span><span class="co">## • step_normalize()</span></span>
<span><span class="co">## • step_pca()</span></span>
<span><span class="co">## • step_normalize()</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ── Model ────────────────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">## Single Layer Neural Network Model Specification (classification)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Main Arguments:</span></span>
<span><span class="co">##   hidden_units = 1</span></span>
<span><span class="co">##   penalty = 1</span></span>
<span><span class="co">##   epochs = 125</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Engine-Specific Arguments:</span></span>
<span><span class="co">##   trace = 0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Computational engine: nnet</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>No more values of <code><a href="https://hardhat.tidymodels.org/reference/tune.html">tune()</a></code> are included in this finalized workflow. Now the model can be fit to the entire training set:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-final-model_1bccdec6958247f0c52e5b371640f0d6">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">final_mlp_fit</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">final_mlp_wflow</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">fit</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This object can now be used to make future predictions on new data.</p>
<p>If you did not use a workflow, finalization of a model and/or recipe is done using <code><a href="https://tune.tidymodels.org/reference/finalize_model.html">finalize_model()</a></code> and <code><a href="https://tune.tidymodels.org/reference/finalize_model.html">finalize_recipe()</a></code>.</p>
</section><section id="sec-tuning-usemodels" class="level2" data-number="13.4"><h2 data-number="13.4" class="anchored" data-anchor-id="sec-tuning-usemodels">
<span class="header-section-number">13.4</span> Tools for Creating Tuning Specifications</h2>
<p>The <span class="pkg">usemodels</span> package can take a data frame and model formula, then write out R code for tuning the model. The code also creates an appropriate recipe whose steps depend on the requested model as well as the predictor data.</p>
<p>For example, for the Ames housing data, <code>xgboost</code> modeling code could be created with:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/models-use-models-code_02a850dbad75b2e30f4435a18961237e">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://usemodels.tidymodels.org/">usemodels</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">use_xgboost</span><span class="op">(</span><span class="va">Sale_Price</span> <span class="op">~</span> <span class="va">Neighborhood</span> <span class="op">+</span> <span class="va">Gr_Liv_Area</span> <span class="op">+</span> <span class="va">Year_Built</span> <span class="op">+</span> <span class="va">Bldg_Type</span> <span class="op">+</span> </span>
<span>              <span class="va">Latitude</span> <span class="op">+</span> <span class="va">Longitude</span>, </span>
<span>            data <span class="op">=</span> <span class="va">ames_train</span>,</span>
<span>            <span class="co"># Add comments explaining some of the code:</span></span>
<span>            verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The resulting code is as follows:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/models-use-models-res_c5028278d06a264eda31f674f515d9e3">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">xgboost_recipe</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">recipe</span><span class="op">(</span>formula <span class="op">=</span> <span class="va">Sale_Price</span> <span class="op">~</span> <span class="va">Neighborhood</span> <span class="op">+</span> <span class="va">Gr_Liv_Area</span> <span class="op">+</span> <span class="va">Year_Built</span> <span class="op">+</span> <span class="va">Bldg_Type</span> <span class="op">+</span> </span>
<span>    <span class="va">Latitude</span> <span class="op">+</span> <span class="va">Longitude</span>, data <span class="op">=</span> <span class="va">ames_train</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">step_novel</span><span class="op">(</span><span class="fu">all_nominal_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="co">## This model requires the predictors to be numeric. The most common </span></span>
<span>  <span class="co">## method to convert qualitative predictors to numeric is to create </span></span>
<span>  <span class="co">## binary indicator variables (aka dummy variables) from these </span></span>
<span>  <span class="co">## predictors. However, for this model, binary indicator variables can be </span></span>
<span>  <span class="co">## made for each of the levels of the factors (known as 'one-hot </span></span>
<span>  <span class="co">## encoding'). </span></span>
<span>  <span class="fu">step_dummy</span><span class="op">(</span><span class="fu">all_nominal_predictors</span><span class="op">(</span><span class="op">)</span>, one_hot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">step_zv</span><span class="op">(</span><span class="fu">all_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">xgboost_spec</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">boost_tree</span><span class="op">(</span>trees <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, min_n <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, tree_depth <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, learn_rate <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, </span>
<span>    loss_reduction <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>, sample_size <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"regression"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"xgboost"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">xgboost_workflow</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">workflow</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">add_recipe</span><span class="op">(</span><span class="va">xgboost_recipe</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">add_model</span><span class="op">(</span><span class="va">xgboost_spec</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">69305</span><span class="op">)</span></span>
<span><span class="va">xgboost_tune</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">tune_grid</span><span class="op">(</span><span class="va">xgboost_workflow</span>, </span>
<span>            resamples <span class="op">=</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="st">"add your rsample object"</span><span class="op">)</span>, </span>
<span>            grid <span class="op">=</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="st">"add number of candidate points"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Based on what <span class="pkg">usemodels</span> understands about the data, this code is the minimal preprocessing required. For other models, operations like <code>step_normalize()</code> are added to fulfill the basic needs of the model. Notice that it is our responsibility, as the modeling practitioner, to choose what <code>resamples</code> to use for tuning, as well as what kind of <code>grid</code>.</p>
<div class="rmdnote">
<p>The <span class="pkg">usemodels</span> package can also be used to create model fitting code with no tuning by setting the argument <code>tune = FALSE</code>.</p>
</div>
</section><section id="sec-efficient-grids" class="level2" data-number="13.5"><h2 data-number="13.5" class="anchored" data-anchor-id="sec-efficient-grids">
<span class="header-section-number">13.5</span> Tools for Efficient Grid Search</h2>
<p>It is possible to make grid search more computationally efficient by applying a few different tricks and optimizations. This section describes several techniques.</p>
<section id="sec-submodel-trick" class="level3" data-number="13.5.1"><h3 data-number="13.5.1" class="anchored" data-anchor-id="sec-submodel-trick">
<span class="header-section-number">13.5.1</span> Submodel optimization</h3>
<p>There are types of models where, from a single model fit, multiple tuning parameters can be evaluated without refitting.</p>
<p>For example, partial least squares (PLS) is a supervised version of principal component analysis <span class="citation" data-cites="Geladi:1986">(<a href="#ref-Geladi:1986" role="doc-biblioref">Geladi and Kowalski 1986</a>)</span>. It creates components that maximize the variation in the predictors (like PCA) but simultaneously tries to maximize the correlation between these predictors and the outcome. We’ll explore PLS more in <a href="#sec-dimensionality">Chapter&nbsp;<span class="quarto-unresolved-ref ref-noprefix">sec-dimensionality</span></a>. One tuning parameter is the number of PLS components to retain. Suppose that a data set with 100 predictors is fit using PLS. The number of possible components to retain can range from one to fifty. However, in many implementations, a single model fit can compute predicted values across many values of <code>num_comp</code>. As a result, a PLS model created with 100 components can also make predictions for any <code>num_comp &lt;= 100</code>. This saves time since, instead of creating redundant model fits, a single fit can be used to evaluate many submodels.</p>
<p>While not all models can exploit this feature, many broadly used ones do.</p>
<ul>
<li><p>Boosting models can typically make predictions across multiple values for the number of boosting iterations.</p></li>
<li><p>Regularization methods, such as the <span class="pkg">glmnet</span> model, can make simultaneous predictions across the amount of regularization used to fit the model.</p></li>
<li><p>Multivariate adaptive regression splines (MARS) adds a set of nonlinear features to linear regression models <span class="citation" data-cites="Friedman:1991p109">(<a href="#ref-Friedman:1991p109" role="doc-biblioref">Friedman 1991</a>)</span>. The number of terms to retain is a tuning parameter, and it is computationally fast to make predictions across many values of this parameter from a single model fit.</p></li>
</ul>
<p>The <span class="pkg">tune</span> package automatically applies this type of optimization whenever an applicable model is tuned.</p>
<p>For example, if a boosted C5.0 classification model <span class="citation" data-cites="apm">(<a href="#ref-apm" role="doc-biblioref">M. Kuhn and Johnson 2013</a>)</span> was fit to the cell data, we can tune the number of boosting iterations (<code>trees</code>). With all other parameters set at their default values, we can evaluate iterations from 1 to 100 on the same resamples as used previously:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-c5_2a1418c1e063f4f629228057af7e776b">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">c5_spec</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">boost_tree</span><span class="op">(</span>trees <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"C5.0"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"classification"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1307</span><span class="op">)</span></span>
<span><span class="va">c5_spec</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">tune_grid</span><span class="op">(</span></span>
<span>    <span class="va">class</span> <span class="op">~</span> <span class="va">.</span>,</span>
<span>    resamples <span class="op">=</span> <span class="va">cell_folds</span>,</span>
<span>    grid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>trees <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    metrics <span class="op">=</span> <span class="va">roc_res</span></span>
<span>  <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Without the submodel optimization, the call to <code><a href="https://tune.tidymodels.org/reference/tune_grid.html">tune_grid()</a></code> used 62.2 minutes to resample 100 submodels. With the optimization, the same call took 100 <em>seconds</em> (a 37-fold speed-up). The reduced time is the difference in <code><a href="https://tune.tidymodels.org/reference/tune_grid.html">tune_grid()</a></code> fitting 1000 models versus 10 models.</p>
<div class="rmdnote">
<p>Even though we fit the model with and without the submodel prediction trick, this optimization is automatically applied by <span class="pkg">parsnip</span>.</p>
</div>
</section><section id="parallel-processing" class="level3" data-number="13.5.2"><h3 data-number="13.5.2" class="anchored" data-anchor-id="parallel-processing">
<span class="header-section-number">13.5.2</span> Parallel processing</h3>
<p>As previously mentioned in <a href="#sec-parallel"><span class="quarto-unresolved-ref">sec-parallel</span></a>, parallel processing is an effective method for decreasing execution time when resampling models. This advantage conveys to model tuning via grid search, although there are additional considerations.</p>
<p>Let’s consider two different parallel processing schemes.</p>
<p>When tuning models via grid search, there are two distinct loops: one over resamples and another over the unique tuning parameter combinations. In pseudocode, this process would look like:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-resamples-algo_e9fc8714e06d311485011bf190af07ea">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">rs</span> <span class="kw">in</span> <span class="va">resamples</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Create analysis and assessment sets</span></span>
<span>  <span class="co"># Preprocess data (e.g. formula or recipe)</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">mod</span> <span class="kw">in</span> <span class="va">configurations</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Fit model {mod} to the {rs} analysis set</span></span>
<span>    <span class="co"># Predict the {rs} assessment set</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default, the <span class="pkg">tune</span> package parallelizes only over resamples (the outer loop), as opposed to both the outer and inner loops.</p>
<p>This is the optimal scenario when the preprocessing method is expensive. However, there are two potential downsides to this approach:</p>
<ul>
<li><p>It limits the achievable speed-ups when the preprocessing is not expensive.</p></li>
<li><p>The number of parallel workers is limited by the number of resamples. For example, with 10-fold cross-validation you can use only 10 parallel workers even when the computer has more than 10 cores.</p></li>
</ul>
<p>To illustrate how the parallel processing works, we’ll use a case where there are 7 model tuning parameter values, with 5-fold cross-validation. <a href="#fig-one-resample-per-worker">Figure&nbsp;<span class="quarto-unresolved-ref">fig-one-resample-per-worker</span></a> shows how the tasks are allocated to the worker processes.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-one-resample-per-worker_fe4d6bcb9100c3fe179eec5e037725d5">
<div class="cell-output-display">
<div id="fig-one-resample-per-worker" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-one-resample-per-worker-1.png" class="img-fluid figure-img" style="width:50.0%" alt="A diagram of the worker processes when parallel processing matches resamples to a specific worker process. After the preprocess operations are finished, each model fit is executed on the same worker process."></p>
<figcaption class="figure-caption">Figure&nbsp;13.5: Worker processes when parallel processing matches resamples to a specific worker process</figcaption></figure>
</div>
</div>
</div>
<p>Note that each fold is assigned to its own worker process and, since only model parameters are being tuned, the preprocessing is conducted once per fold/worker. If fewer than five worker processes were used, some workers would receive multiple folds.</p>
<p>In the control functions for the <code>tune_*()</code> functions, the argument <code>parallel_over</code> controls how the process is executed. To use the previous parallelization strategy, the argument is <code>parallel_over = "resamples"</code>.</p>
<p>Instead of parallel processing the resamples, an alternate scheme combines the loops over resamples and models into a single loop. In pseudocode, this process would look like:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-everything-algo_f8c21c522d798c3e5ca18b99e50a8b54">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">all_tasks</span> <span class="op">&lt;-</span> <span class="fu">crossing</span><span class="op">(</span><span class="va">resamples</span>, <span class="va">configurations</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">iter</span> <span class="kw">in</span> <span class="va">all_tasks</span><span class="op">)</span> <span class="op">{</span>                           </span>
<span>  <span class="co"># Create analysis and assessment sets for {iter}</span></span>
<span>  <span class="co"># Preprocess data (e.g. formula or recipe)</span></span>
<span>  <span class="co"># Fit model {iter} to the {iter} analysis set</span></span>
<span>  <span class="co"># Predict the {iter} assessment set</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this case, parallelization now occurs over the single loop. For example, if we use 5-fold cross-validation with <span class="math inline">\(M\)</span> tuning parameter values, the loop is executed over <span class="math inline">\(5\times M\)</span> iterations. This increases the number of potential workers that can be used. However, the work related to data preprocessing is repeated multiple times. If those steps are expensive, this approach will be inefficient.</p>
<p>In tidymodels, validation sets are treated as a single resample. In these cases, this parallelization scheme would be best.</p>
<p><a href="#fig-distributed-tasks">Figure&nbsp;<span class="quarto-unresolved-ref">fig-distributed-tasks</span></a> illustrates the delegation of tasks to the workers in this scheme; the same example is used but with 10 workers.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-distributed-tasks_c3b622c1c17e8bc0fc2004103aa9fa19">
<div class="cell-output-display">
<div id="fig-distributed-tasks" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-distributed-tasks-1.png" class="img-fluid figure-img" style="width:70.0%" alt="A diagram of the worker processes when preprocessing and modeling tasks are distributed to many workers. In this instance, more comprehensive parallelization is used but some preprocessing tasks are repeated across worker processes."></p>
<figcaption class="figure-caption">Figure&nbsp;13.6: Worker processes when preprocessing and modeling tasks are distributed to many workers</figcaption></figure>
</div>
</div>
</div>
<p>Here, each worker process handles multiple folds, and the preprocessing is needlessly repeated. For example, for the first fold, the preprocessing was computed seven times instead of once.</p>
<p>For this scheme, the control function argument is <code>parallel_over = "everything"</code>.</p>
</section><section id="benchmarking-boosted-trees" class="level3" data-number="13.5.3"><h3 data-number="13.5.3" class="anchored" data-anchor-id="benchmarking-boosted-trees">
<span class="header-section-number">13.5.3</span> Benchmarking boosted trees</h3>
<p>To compare different possible parallelization schemes, we tuned a boosted tree with the <span class="pkg">xgboost</span> engine using a data set of 4,000 samples, with 5-fold cross-validation and 10 candidate models. These data required some baseline preprocessing that did not require any estimation. The preprocessing was handled three different ways:</p>
<ol type="1">
<li>Preprocess the data prior to modeling using a <span class="pkg">dplyr</span> pipeline (labeled as “none” in the later plots).</li>
<li>Conduct the same preprocessing via a recipe (shown as “light” preprocessing).</li>
<li>With a recipe, add an additional step that has a high computational cost (labeled as “expensive”).</li>
</ol>
<p>The first and second preprocessing options are designed for comparison, to measure the computational cost of the recipe in the second option. The third option measures the cost of performing redundant computations with <code>parallel_over = "everything"</code>.</p>
<p>We evaluated this process using variable numbers of worker processes and using the two <code>parallel_over</code> options, on a computer with 10 physical cores and 20 virtual cores (via hyper-threading).</p>
<p>First, let’s consider the raw execution times in <a href="#fig-parallel-times">Figure&nbsp;<span class="quarto-unresolved-ref">fig-parallel-times</span></a>.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-parallel-times_7e808c8b2d3748de2a4045a525d9e0be">
<div class="cell-output-display">
<div id="fig-parallel-times" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-parallel-times-1.png" class="img-fluid figure-img" style="width:70.0%" alt="Execution times for model tuning versus the number of workers using different delegation schemes."></p>
<figcaption class="figure-caption">Figure&nbsp;13.7: Execution times for model tuning versus the number of workers using different delegation schemes</figcaption></figure>
</div>
</div>
</div>
<p>Since there were only five resamples, the number of cores used when <code>parallel_over = "resamples"</code> is limited to five.</p>
<p>Comparing the curves in the first two panels for “none” and “light”:</p>
<ul>
<li><p>There is little difference in the execution times between the panels. This indicates, for these data, there is no real computational penalty for doing the preprocessing steps in a recipe.</p></li>
<li><p>There is some benefit for using <code>parallel_over = "everything"</code> with many cores. However, as shown in the figure, the majority of the benefit of parallel processing occurs in the first five workers.</p></li>
</ul>
<p>With the expensive preprocessing step, there is a considerable difference in execution times. Using <code>parallel_over = "everything"</code> is problematic since, even using all cores, it never achieves the execution time that <code>parallel_over = "resamples"</code> attains with just five cores. This is because the costly preprocessing step is unnecessarily repeated in the computational scheme.</p>
<p>We can also view these data in terms of speed-ups in <a href="#fig-parallel-speedups">Figure&nbsp;<span class="quarto-unresolved-ref">fig-parallel-speedups</span></a>.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-parallel-speedups_169ef2d469ff859059e3000e58c7b940">
<div class="cell-output-display">
<div id="fig-parallel-speedups" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-parallel-speedups-1.png" class="img-fluid figure-img" style="width:70.0%" alt="Speed-ups for model tuning versus the number of workers using different delegation schemes. The diagonal black line indicates a linear speedup where the addition of a new worker process has maximal effect. The 'everything' scheme shows that the benefits decrease after three or four workers, especially when there is expensive preprocessing. The 'resamples' scheme has almost linear speedups across all tasks."></p>
<figcaption class="figure-caption">Figure&nbsp;13.8: Speed-ups for model tuning versus the number of workers using different delegation schemes. The diagonal black line indicates a linear speedup where the addition of a new worker process has maximal effect.</figcaption></figure>
</div>
</div>
</div>
<p>The best speed-ups, for these data, occur when <code>parallel_over = "resamples"</code> and when the computations are expensive. However, in the latter case, remember that the previous analysis indicates that the overall model fits are slower.</p>
<p>What is the benefit of using the submodel optimization method in conjunction with parallel processing? The C5.0 classification model shown in <a href="#sec-submodel-trick"><span class="quarto-unresolved-ref">sec-submodel-trick</span></a> was also run in parallel with ten workers. The parallel computations took 13.3 seconds for a 7.5-fold speed-up (both runs used the submodel optimization trick). Between the submodel optimization trick and parallel processing, there was a total 282-fold speed-up over the most basic grid search code.</p>
<div class="rmdwarning">
<p>Overall, note that the increased computational savings will vary from model to model and are also affected by the size of the grid, the number of resamples, etc. A very computationally efficient model may not benefit as much from parallel processing.</p>
</div>
</section><section id="access-to-global-variables" class="level3" data-number="13.5.4"><h3 data-number="13.5.4" class="anchored" data-anchor-id="access-to-global-variables">
<span class="header-section-number">13.5.4</span> Access to global variables</h3>
<p>When using tidymodels, it is possible to use values in your local environment (usually the global environment) in model objects.</p>
<div class="rmdnote">
<p>What do we mean by “environment” here? Think of an environment in R as a place to store variables that you can work with. See the “Environments” chapter of <span class="citation" data-cites="wickham2019advanced">Wickham (<a href="#ref-wickham2019advanced" role="doc-biblioref">2019</a>)</span> to learn more.</p>
</div>
<p>If we define a variable to use as a model parameter and then pass it to a function like <code>linear_reg()</code>, the variable is typically defined in the global environment.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-spec-global_c4545922d996ccfeb73a2cd4c45ae40a">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">coef_penalty</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">spec</span> <span class="op">&lt;-</span> <span class="fu">linear_reg</span><span class="op">(</span>penalty <span class="op">=</span> <span class="va">coef_penalty</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">set_engine</span><span class="op">(</span><span class="st">"glmnet"</span><span class="op">)</span></span>
<span><span class="va">spec</span></span>
<span><span class="co">## Linear Regression Model Specification (regression)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Main Arguments:</span></span>
<span><span class="co">##   penalty = coef_penalty</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Computational engine: glmnet</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Models created with the parsnip package save arguments like these as <em>quosures</em>; these are objects that track both the name of the object as well as the environment where it lives:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-spec-quo_0cfa0644a2d936a278f624a7665bf973">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spec</span><span class="op">$</span><span class="va">args</span><span class="op">$</span><span class="va">penalty</span></span>
<span><span class="co">## &lt;quosure&gt;</span></span>
<span><span class="co">## expr: ^coef_penalty</span></span>
<span><span class="co">## env:  global</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Notice that we have <code>env:  global</code> because this variable was created in the global environment. The model specification defined by <code>spec</code> works correctly when run in a user’s regular session because that session is also using the global environment; R can easily find the object <code>coef_penalty</code>.</p>
<div class="rmdwarning">
<p>When such a model is evaluated with parallel workers, it may fail. Depending on the particular technology that is used for parallel processing, the workers may not have access to the global environment.</p>
</div>
<p>When writing code that will be run in parallel, it is a good idea to insert the actual data into the objects rather than the reference to the object. The <span class="pkg">rlang</span> and <span class="pkg">dplyr</span> packages can be very helpful for this. For example, the <code>!!</code> operator can splice a single value into an object:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-spec-bang-bang_56e7ab473d44998cc9acb09b1b4d6057">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spec</span> <span class="op">&lt;-</span> <span class="fu">linear_reg</span><span class="op">(</span>penalty <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="va">coef_penalty</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">set_engine</span><span class="op">(</span><span class="st">"glmnet"</span><span class="op">)</span></span>
<span><span class="va">spec</span><span class="op">$</span><span class="va">args</span><span class="op">$</span><span class="va">penalty</span></span>
<span><span class="co">## &lt;quosure&gt;</span></span>
<span><span class="co">## expr: ^0.1</span></span>
<span><span class="co">## env:  empty</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now the output is <code>^0.1</code>, indicating that the value is there instead of the reference to the object. When you have multiple external values to insert into an object, the <code>!!!</code> operator can help:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-spec-bang-bang-bang_6a5664a0088475ff6838a3c4f503a102">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mcmc_args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>chains <span class="op">=</span> <span class="fl">3</span>, iter <span class="op">=</span> <span class="fl">1000</span>, cores <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">linear_reg</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">set_engine</span><span class="op">(</span><span class="st">"stan"</span>, <span class="op">!</span><span class="op">!</span><span class="op">!</span><span class="va">mcmc_args</span><span class="op">)</span></span>
<span><span class="co">## Linear Regression Model Specification (regression)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Engine-Specific Arguments:</span></span>
<span><span class="co">##   chains = 3</span></span>
<span><span class="co">##   iter = 1000</span></span>
<span><span class="co">##   cores = 3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Computational engine: stan</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Recipe selectors are another place where you might want access to global variables. Suppose you have a recipe step that should use all of the predictors in the cell data that were measured using the second optical channel. We can create a vector of these column names:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-ch-2_1fec5a5f1767f1fff3cc35e7743fa348">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="va">ch_2_vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html">str_subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, <span class="st">"ch_2"</span><span class="op">)</span></span>
<span><span class="va">ch_2_vars</span></span>
<span><span class="co">## [1] "avg_inten_ch_2"   "total_inten_ch_2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We could hard-code these into a recipe step but it would be better to reference them programmatically in case the data change. Two ways to do this are:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-ch-2-rec_c6cd7edb738155e9b294b4c5b82db1f7">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Still uses a reference to global data (~_~;)</span></span>
<span><span class="fu">recipe</span><span class="op">(</span><span class="va">class</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">cells</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">step_spatialsign</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">ch_2_vars</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ── Recipe ───────────────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ── Inputs</span></span>
<span><span class="co">## Number of variables by role</span></span>
<span><span class="co">## outcome:    1</span></span>
<span><span class="co">## predictor: 56</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ── Operations</span></span>
<span><span class="co">## • Spatial sign on: all_of(ch_2_vars)</span></span>
<span></span>
<span><span class="co"># Inserts the values into the step ヽ(•‿•)ノ</span></span>
<span><span class="fu">recipe</span><span class="op">(</span><span class="va">class</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">cells</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">step_spatialsign</span><span class="op">(</span><span class="op">!</span><span class="op">!</span><span class="op">!</span><span class="va">ch_2_vars</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ── Recipe ───────────────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ── Inputs</span></span>
<span><span class="co">## Number of variables by role</span></span>
<span><span class="co">## outcome:    1</span></span>
<span><span class="co">## predictor: 56</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ── Operations</span></span>
<span><span class="co">## • Spatial sign on: "avg_inten_ch_2", "total_inten_ch_2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The latter is better for parallel processing because all of the needed information is embedded in the recipe object.</p>
</section><section id="sec-racing" class="level3" data-number="13.5.5"><h3 data-number="13.5.5" class="anchored" data-anchor-id="sec-racing">
<span class="header-section-number">13.5.5</span> Racing methods</h3>
<p>One issue with grid search is that all models need to be fit across all resamples before any tuning parameters can be evaluated. It would be helpful if instead, at some point during tuning, an interim analysis could be conducted to eliminate any truly awful parameter candidates. This would be akin to <em>futility analysis</em> in clinical trials. If a new drug is performing excessively poorly (or well), it is potentially unethical to wait until the trial finishes to make a decision.</p>
<p>In machine learning, the set of techniques called <em>racing methods</em> provide a similar function <span class="citation" data-cites="maron1994hoeffding">(<a href="#ref-maron1994hoeffding" role="doc-biblioref">Maron and Moore 1994</a>)</span>. Here, the tuning process evaluates all models on an initial subset of resamples. Based on their current performance metrics, some parameter sets are not considered in subsequent resamples.</p>
<p>As an example, in the multilayer perceptron tuning process with a regular grid explored in this chapter, what would the results look like after only the first three folds? Using techniques similar to those shown in <a href="#sec-compare">Chapter&nbsp;<span class="quarto-unresolved-ref ref-noprefix">sec-compare</span></a>, we can fit a model where the outcome is the resampled area under the ROC curve and the predictor is an indicator for the parameter combination. The model takes the resample-to-resample effect into account and produces point and interval estimates for each parameter setting. The results of the model are one-sided 95% confidence intervals that measure the loss of the ROC value relative to the currently best performing parameters, as shown in <a href="#fig-racing-process">Figure&nbsp;<span class="quarto-unresolved-ref">fig-racing-process</span></a>.</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/fig-racing-process_61e85bb876ed608664abb87ca140ff42">
<div class="cell-output-display">
<div id="fig-racing-process" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="13-grid-search_files/figure-html/fig-racing-process-1.png" class="img-fluid figure-img" style="width:80.0%" alt="An illustration of the racing process for 20 tuning parameters and 10 resamples. The analysis is conducted at the first, third, and last resample. As the number of resamples increases, the confidence intervals show some model configurations that do not have confidence intervals that overlap with zero. These are excluded from subsequent resamples."></p>
<figcaption class="figure-caption">Figure&nbsp;13.9: The racing process for 20 tuning parameters and 10 resamples</figcaption></figure>
</div>
</div>
</div>
<p>Any parameter set whose confidence interval includes zero would lack evidence that its performance is statistically different from the best results. We retain 6 settings; these are resampled more. The remaining 14 submodels are no longer considered.</p>
<video width="720" height="720" controls=""><source src="race_results.mp4" type="video/mp4"></video><p>The process continues for each resample; after the next set of performance metrics, a new model is fit to these statistics, and more submodels are potentially discarded.<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></p>
<div class="rmdwarning">
<p>Racing methods can be more efficient than basic grid search as long as the interim analysis is fast and some parameter settings have poor performance. It also is most helpful when the model does <em>not</em> have the ability to exploit submodel predictions.</p>
</div>
<p>The <span class="pkg">finetune</span> package contains functions for racing. The <code><a href="https://finetune.tidymodels.org/reference/tune_race_anova.html">tune_race_anova()</a></code> function conducts an ANOVA model to test for statistical significance of the different model configurations. The syntax to reproduce the filtering shown previously is:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-cells-race-code_29d57aa1fa57f6aaec1136d0a1b59ef8">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidymodels/finetune">finetune</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1308</span><span class="op">)</span></span>
<span><span class="va">mlp_sfd_race</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">mlp_wflow</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://finetune.tidymodels.org/reference/tune_race_anova.html">tune_race_anova</a></span><span class="op">(</span></span>
<span>    <span class="va">cell_folds</span>,</span>
<span>    grid <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    param_info <span class="op">=</span> <span class="va">mlp_param</span>,</span>
<span>    metrics <span class="op">=</span> <span class="va">roc_res</span>,</span>
<span>    control <span class="op">=</span> <span class="fu"><a href="https://finetune.tidymodels.org/reference/control_race.html">control_race</a></span><span class="op">(</span>verbose_elim <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The arguments mirror those of <code><a href="https://tune.tidymodels.org/reference/tune_grid.html">tune_grid()</a></code>. The function <code><a href="https://finetune.tidymodels.org/reference/control_race.html">control_race()</a></code> has options for the elimination procedure.</p>
<p>As shown in the animation above, there were one tuning parameter combinations under consideration once the full set of resamples were evaluated. <code><a href="https://tune.tidymodels.org/reference/show_best.html">show_best()</a></code> returns the best models (ranked by performance) but returns only the configurations that were never eliminated:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/grid-race-best_76366ec3d832d12f3e1c050e91a633c0">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://tune.tidymodels.org/reference/show_best.html">show_best</a></span><span class="op">(</span><span class="va">mlp_sfd_race</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 10</span></span>
<span><span class="co">##   hidden_units penalty epochs num_comp .metric .estimator  mean     n std_err</span></span>
<span><span class="co">##          &lt;int&gt;   &lt;dbl&gt;  &lt;int&gt;    &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;      &lt;dbl&gt; &lt;int&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">## 1            8   0.814    177       15 roc_auc binary     0.890    10 0.00966</span></span>
<span><span class="co">## # ℹ 1 more variable: .config &lt;chr&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are other interim analysis techniques for discarding settings. For example, <span class="citation" data-cites="krueger15a">Krueger, Panknin, and Braun (<a href="#ref-krueger15a" role="doc-biblioref">2015</a>)</span> use traditional sequential analysis methods whereas <span class="citation" data-cites="kuhn2014futility">Max Kuhn (<a href="#ref-kuhn2014futility" role="doc-biblioref">2014</a>)</span> treats the data as a sports competition and uses the Bradley-Terry model <span class="citation" data-cites="bradley1952rank">(<a href="#ref-bradley1952rank" role="doc-biblioref">Bradley and Terry 1952</a>)</span> to measure the winning ability of parameter settings.</p>
</section></section><section id="sec-grid-summary" class="level2" data-number="13.6"><h2 data-number="13.6" class="anchored" data-anchor-id="sec-grid-summary">
<span class="header-section-number">13.6</span> Chapter Summary</h2>
<p>This chapter discussed the two main classes of grid search (regular and non-regular) that can be used for model tuning and demonstrated how to construct these grids, either manually or using the family of <code>grid_*()</code> functions. The <code><a href="https://tune.tidymodels.org/reference/tune_grid.html">tune_grid()</a></code> function can evaluate these candidate sets of model parameters using resampling. The chapter also showed how to finalize a model, recipe, or workflow to update the parameter values for the final fit. Grid search can be computationally expensive, but thoughtful choices in the experimental design of such searches can make them tractable.</p>
<p>The data analysis code that will be reused in the next chapter is:</p>
<div class="cell" data-layout-align="center" data-hash="13-grid-search_cache/html/resampling-summary_cff382e9ff11c5712a1a11fcc70960f0">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidymodels.tidymodels.org">tidymodels</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="va">cells</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">case</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1304</span><span class="op">)</span></span>
<span><span class="va">cell_folds</span> <span class="op">&lt;-</span> <span class="fu">vfold_cv</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span></span>
<span><span class="va">roc_res</span> <span class="op">&lt;-</span> <span class="fu">metric_set</span><span class="op">(</span><span class="va">roc_auc</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-BHH" class="csl-entry" role="listitem">
Box, GEP, W Hunter, and J Hunter. 2005. <em>Statistics for Experimenters: An Introduction to Design, Data Analysis, and Model Building</em>. Wiley.
</div>
<div id="ref-bradley1952rank" class="csl-entry" role="listitem">
Bradley, R, and M Terry. 1952. <span>“Rank Analysis of Incomplete Block Designs: <span>I.</span> The Method of Paired Comparisons.”</span> <em>Biometrika</em> 39 (3/4): 324–45.
</div>
<div id="ref-Friedman:1991p109" class="csl-entry" role="listitem">
Friedman, J. 1991. <span>“Multivariate Adaptive Regression Splines.”</span> <em>The Annals of Statistics</em> 19 (1): 1–141.
</div>
<div id="ref-Geladi:1986" class="csl-entry" role="listitem">
Geladi, P., and B Kowalski. 1986. <span>“Partial Least-Squares Regression: A Tutorial.”</span> <em>Analytica Chimica Acta</em> 185: 1–17.
</div>
<div id="ref-Hill" class="csl-entry" role="listitem">
Hill, A, P LaPan, Y Li, and S Haney. 2007. <span>“Impact of Image Segmentation on High-Content Screening Data Quality for <span>SK</span>-<span>BR</span>-3 Cells.”</span> <em>BMC Bioinformatics</em> 8 (1): 340.
</div>
<div id="ref-maxproj" class="csl-entry" role="listitem">
Joseph, V, E Gul, and S Ba. 2015. <span>“Maximum Projection Designs for Computer Experiments.”</span> <em>Biometrika</em> 102 (2): 371–80.
</div>
<div id="ref-krueger15a" class="csl-entry" role="listitem">
Krueger, T, D Panknin, and M Braun. 2015. <span>“Fast Cross-Validation via Sequential Testing.”</span> <em>Journal of Machine Learning Research</em> 16 (33): 1103–55.
</div>
<div id="ref-kuhn2014futility" class="csl-entry" role="listitem">
Kuhn, Max. 2014. <span>“Futility Analysis in the Cross-Validation of Machine Learning Models.”</span> <a href="https://arxiv.org/abs/1405.6974">https://arxiv.org/abs/1405.6974</a>.
</div>
<div id="ref-apm" class="csl-entry" role="listitem">
Kuhn, M, and K Johnson. 2013. <em>Applied Predictive Modeling</em>. Springer.
</div>
<div id="ref-maron1994hoeffding" class="csl-entry" role="listitem">
Maron, O, and A Moore. 1994. <span>“Hoeffding Races: Accelerating Model Selection Search for Classification and Function Approximation.”</span> In <em>Advances in Neural Information Processing Systems</em>, 59–66.
</div>
<div id="ref-lhd" class="csl-entry" role="listitem">
McKay, M, R Beckman, and W Conover. 1979. <span>“A Comparison of Three Methods for Selecting Values of Input Variables in the Analysis of Output from a Computer Code.”</span> <em>Technometrics</em> 21 (2): 239–45.
</div>
<div id="ref-santner2003design" class="csl-entry" role="listitem">
Santner, T, B Williams, W Notz, and B Williams. 2003. <em>The Design and Analysis of Computer Experiments</em>. Springer.
</div>
<div id="ref-maxent" class="csl-entry" role="listitem">
Shewry, M, and H Wynn. 1987. <span>“Maximum Entropy Sampling.”</span> <em>Journal of Applied Statistics</em> 14 (2): 165–70.
</div>
<div id="ref-wickham2019advanced" class="csl-entry" role="listitem">
Wickham, H. 2019. <em>Advanced r</em>. 2nd ed. Chapman &amp; Hall/CRC the r Series. Taylor &amp; Francis. <a href="https://doi.org/10.1201/9781351201315">https://doi.org/10.1201/9781351201315</a>.
</div>
<div id="ref-yeo2000new" class="csl-entry" role="listitem">
Yeo, I-K, and R Johnson. 2000. <span>“A New Family of Power Transformations to Improve Normality or Symmetry.”</span> <em>Biometrika</em> 87 (4): 954–59.
</div>
</div>
</section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><hr>
<ol>
<li id="fn1"><p><a href="https://CRAN.R-project.org/view=ExperimentalDesign" class="uri">https://CRAN.R-project.org/view=ExperimentalDesign</a><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>See <span class="citation" data-cites="kuhn2014futility">Max Kuhn (<a href="#ref-kuhn2014futility" role="doc-biblioref">2014</a>)</span> for more details on the computational aspects of this approach.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./12-tuning-parameters.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Model Tuning and the Dangers of Overfitting</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./14-iterative-search.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Iterative Search</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Modelado Ordenado con R fue escrito por Max Kuhn, y Julia Silge.</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Este libro fue creado usando <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>


</body></html>