<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Max Kuhn, y Julia Silge">
<title>Modelado Ordenado con R - 21&nbsp; Inferential Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./pre-proc-table.html" rel="next">
<link href="./20-ensemble-models.html" rel="prev">
<link href="./images/cover.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><link rel="stylesheet" href="style.css">
<link rel="stylesheet" href="TMwR.css">
</head>
<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./16-dimensionality-reduction.html">BEYOND THE BASICS</a></li><li class="breadcrumb-item"><a href="./21-inferential-analysis.html"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Inferential Analysis</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Modelado Ordenado con R</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/davidrsch/TMwRes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hello World</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">INTRODUCTION</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-software-modeling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Software for modeling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">A Tidyverse Primer</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-base-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">A Review of R Modeling Fundamentals</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">MODELING BASICS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-ames.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">The Ames Housing Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-data-spending.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Spending our Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-fitting-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Fitting Models with parsnip</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-the-model-workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">A Model Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-feature-engineering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Feature Engineering with recipes</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-judging-model-effectiveness.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Judging Model Effectiveness</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">TOOLS FOR CREATING EFFECTIVE MODELS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-resampling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Resampling for Evaluating Performance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-comparing-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Comparing Models with Resampling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12-tuning-parameters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Model Tuning and the Dangers of Overfitting</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13-grid-search.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Grid Search</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14-iterative-search.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Iterative Search</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./15-workflow-sets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Screening Many Models</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">BEYOND THE BASICS</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./16-dimensionality-reduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Dimensionality Reduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./17-encoding-categorical-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Encoding Categorical Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./18-explaining-models-and-predictions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Explaining Models and Predictions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./19-when-should-you-trust-predictions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">When Should You Trust Your Predictions?</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./20-ensemble-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Ensembles of Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21-inferential-analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Inferential Analysis</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pre-proc-table.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Recommended Preprocessing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#inference-for-count-data" id="toc-inference-for-count-data" class="nav-link active" data-scroll-target="#inference-for-count-data"><span class="header-section-number">21.1</span> Inference for Count Data</a></li>
  <li><a href="#comparisons-with-two-sample-tests" id="toc-comparisons-with-two-sample-tests" class="nav-link" data-scroll-target="#comparisons-with-two-sample-tests"><span class="header-section-number">21.2</span> Comparisons with Two-Sample Tests</a></li>
  <li><a href="#log-linear-models" id="toc-log-linear-models" class="nav-link" data-scroll-target="#log-linear-models"><span class="header-section-number">21.3</span> Log-Linear Models</a></li>
  <li><a href="#a-more-complex-model" id="toc-a-more-complex-model" class="nav-link" data-scroll-target="#a-more-complex-model"><span class="header-section-number">21.4</span> A More Complex Model</a></li>
  <li><a href="#inference-options" id="toc-inference-options" class="nav-link" data-scroll-target="#inference-options"><span class="header-section-number">21.5</span> More Inferential Analysis</a></li>
  <li><a href="#inference-summary" id="toc-inference-summary" class="nav-link" data-scroll-target="#inference-summary"><span class="header-section-number">21.6</span> Chapter Summary</a></li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/davidrsch/TMwRes/edit/main/21-inferential-analysis.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/davidrsch/TMwRes/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="inferential" class="quarto-section-identifier"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Inferential Analysis</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="rmdnote">
<p>In Section @ref(model-types), we outlined a taxonomy of models and said that most models can be categorized as descriptive, inferential, and/or predictive.</p>
</div>
<p>Most of the chapters in this book have focused on models from the perspective of the accuracy of predicted values, an important quality of models for all purposes but most relevant for predictive models. Inferential models are usually created not only for their predictions, but also to make inferences or judgments about some component of the model, such as a coefficient value or other parameter. These results are often used to answer some (hopefully) pre-defined questions or hypotheses. In predictive models, predictions on hold-out data are used to validate or characterize the quality of the model. Inferential methods focus on validating the probabilistic or structural assumptions that are made prior to fitting the model.</p>
<p>For example, in ordinary linear regression, the common assumption is that the residual values are independent and follow a Gaussian distribution with a constant variance. While you may have scientific or domain knowledge to lend credence to this assumption for your model analysis, the residuals from the fitted model are usually examined to determine if the assumption was a good idea. As a result, the methods for determining if the model’s assumptions have been met are not as simple as looking at holdout predictions, although that can be very useful as well.</p>
<p>We will use p-values in this chapter. However, the tidymodels framework tends to promote confidence intervals over p-values as a method for quantifying the evidence for an alternative hypothesis. As previously shown in Section @ref(tidyposterior), Bayesian methods are often superior to both p-values and confidence intervals in terms of ease of interpretation (but they can be more computationally expensive).</p>
<div class="rmdwarning">
<p>There has been a push in recent years to move away from p-values in favor of other methods <span class="citation" data-cites="pvalue">(<a href="references.html#ref-pvalue" role="doc-biblioref">Wasserstein and Lazar 2016</a>)</span>. See Volume 73 of <a href="https://www.tandfonline.com/toc/utas20/73/"><em>The American Statistician</em></a> for more information and discussion.</p>
</div>
<p>In this chapter, we describe how to use <span class="pkg">tidymodels</span> for fitting and assessing inferential models. In some cases, the tidymodels framework can help users work with the objects produced by their models. In others, it can help assess the quality of a given model.</p>
<section id="inference-for-count-data" class="level2" data-number="21.1"><h2 data-number="21.1" class="anchored" data-anchor-id="inference-for-count-data">
<span class="header-section-number">21.1</span> Inference for Count Data</h2>
<p>To understand how tidymodels packages can be used for inferential modeling, let’s focus on an example with count data. We’ll use biochemistry publication data from the <span class="pkg">pscl</span> package. These data consist of information on 915 Ph.D.&nbsp;biochemistry graduates and tries to explain factors that impact their academic productivity (measured via number or count of articles published within three years). The predictors include the gender of the graduate, their marital status, the number of children of the graduate that are at least five years old, the prestige of their department, and the number of articles produced by their mentor in the same time period. The data reflect biochemistry doctorates who finished their education between 1956 and 1963. The data are a somewhat biased sample of all of the biochemistry doctorates given during this period (based on completeness of information).</p>
<div class="rmdnote">
<p>Recall that in Chapter @ref(trust) we asked the question “Is our model applicable for predicting a specific data point?” It is very important to define what populations an inferential analysis applies to. For these data, the results would likely apply to biochemistry doctorates given around the time frame that the data were collected. Does it also apply to other chemistry doctorate types (e.g., medicinal chemistry, etc)? These are important questions to address (and document) when conducting inferential analyses.</p>
</div>
<p>A plot of the data shown in Figure @ref(fig:counts) indicates that many graduates did not publish any articles in this time and that the outcome follows a right-skewed distribution:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidymodels.tidymodels.org">tidymodels</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://tidymodels.tidymodels.org/reference/tidymodels_prefer.html">tidymodels_prefer</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"bioChemists"</span>, package <span class="op">=</span> <span class="st">"pscl"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">bioChemists</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">art</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Number of articles within 3y of graduation"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="figures/counts-1.png" class="img-fluid figure-img" style="width:80.0%" alt="The distribution of the number of articles written within 3 years of graduation. The distribution is right-skewed and most of the data have counts of zero or one."></p>
<figcaption class="figure-caption">Distribution of the number of articles written within 3 years of graduation</figcaption></figure>
</div>
</div>
</div>
<p>Since the outcome data are counts, the most common distribution assumption to make is that the outcome has a Poisson distribution. This chapter will use these data for several types of analyses.</p>
</section><section id="comparisons-with-two-sample-tests" class="level2" data-number="21.2"><h2 data-number="21.2" class="anchored" data-anchor-id="comparisons-with-two-sample-tests">
<span class="header-section-number">21.2</span> Comparisons with Two-Sample Tests</h2>
<p>We can start with hypothesis testing. The original author’s goal with this data set on biochemistry publication data was to determine if there is a difference in publications between men and women <span class="citation" data-cites="Long1992">(<a href="references.html#ref-Long1992" role="doc-biblioref">Long 1992</a>)</span>. The data from the study show:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">bioChemists</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">fem</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">summarize</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">art</span><span class="op">)</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">art</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 2 × 3</span></span>
<span><span class="co">##   fem   counts     n</span></span>
<span><span class="co">##   &lt;fct&gt;  &lt;int&gt; &lt;int&gt;</span></span>
<span><span class="co">## 1 Men      930   494</span></span>
<span><span class="co">## 2 Women    619   421</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There were many more publications by men, although there were also more men in the data. The simplest approach to analyzing these data would be to do a two-sample comparison using the <code><a href="https://rdrr.io/r/stats/poisson.test.html">poisson.test()</a></code> function in the <span class="pkg">stats</span> package. It requires the counts for one or two groups.</p>
<p>For our application, the hypotheses to compare the two sexes are:</p>
<span class="math display">\[\begin{align}
H_0&amp;: \lambda_m = \lambda_f \notag \\
H_a&amp;: \lambda_m \ne \lambda_f \notag
\end{align}\]</span>
<p>where the <span class="math inline">\(\lambda\)</span> values are the rates of publications (over the same time period).</p>
<p>A basic application of the test is:<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/poisson.test.html">poisson.test</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">930</span>, <span class="fl">619</span><span class="op">)</span>, T <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Comparison of Poisson rates</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## data:  c(930, 619) time base: 3</span></span>
<span><span class="co">## count1 = 930, expected count1 = 774, p-value = 3e-15</span></span>
<span><span class="co">## alternative hypothesis: true rate ratio is not equal to 1</span></span>
<span><span class="co">## 95 percent confidence interval:</span></span>
<span><span class="co">##  1.356 1.666</span></span>
<span><span class="co">## sample estimates:</span></span>
<span><span class="co">## rate ratio </span></span>
<span><span class="co">##      1.502</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function reports a p-value as well as a confidence interval for the ratio of the publication rates. The results indicate that the observed difference is greater than the experiential noise and favors <span class="math inline">\(H_a\)</span>.</p>
<p>One issue with using this function is that the results come back as an <code>htest</code> object. While this type of object has a well-defined structure, it can be difficult to consume for subsequent operations such as reporting or visualizations. The most impactful tool that tidymodels offers for inferential models is the <code><a href="https://generics.r-lib.org/reference/tidy.html">tidy()</a></code> functions in the <span class="pkg">broom</span> package. As previously seen, this function makes a well-formed, predictably named tibble from the object. We can <code><a href="https://generics.r-lib.org/reference/tidy.html">tidy()</a></code> the results of our two-sample comparison test:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/poisson.test.html">poisson.test</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">930</span>, <span class="fl">619</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">tidy</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 8</span></span>
<span><span class="co">##   estimate statistic  p.value parameter conf.low conf.high method        alternative</span></span>
<span><span class="co">##      &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;      </span></span>
<span><span class="co">## 1     1.50       930 2.73e-15      774.     1.36      1.67 Comparison o… two.sided</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="rmdnote">
<p>Between the <a href="https://broom.tidymodels.org/"><span class="pkg">broom</span></a> and <a href="https://CRAN.R-project.org/package=broom.mixed"><span class="pkg">broom.mixed</span></a> packages, there are <code><a href="https://generics.r-lib.org/reference/tidy.html">tidy()</a></code> methods for more than 150 models.</p>
</div>
<p>While the Poisson distribution is reasonable, we might also want to assess using fewer distributional assumptions. Two methods that might be helpful are the bootstrap and permutation tests <span class="citation" data-cites="davison1997bootstrap">(<a href="references.html#ref-davison1997bootstrap" role="doc-biblioref">Davison and Hinkley 1997</a>)</span>.</p>
<p>The <span class="pkg">infer</span> package, part of the tidymodels framework, is a powerful and intuitive tool for hypothesis testing <span class="citation" data-cites="ModernDive">(<a href="references.html#ref-ModernDive" role="doc-biblioref">Ismay and Kim 2021</a>)</span>. Its syntax is concise and designed for nonstatisticians.</p>
<p>First, we <code><a href="https://infer.tidymodels.org/reference/specify.html">specify()</a></code> that we will use the difference in the mean number of articles between the sexes and then <code><a href="https://infer.tidymodels.org/reference/calculate.html">calculate()</a></code> the statistic from the data. Recall that the maximum likelihood estimator for the Poisson mean is the sample mean. The hypotheses tested here are the same as the previous test (but are conducted using a different testing procedure).</p>
<p>With <span class="pkg">infer</span>, we specify the outcome and covariate, then state the statistic of interest:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidymodels/infer">infer</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">bioChemists</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/specify.html">specify</a></span><span class="op">(</span><span class="va">art</span> <span class="op">~</span> <span class="va">fem</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/calculate.html">calculate</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"diff in means"</span>, order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Men"</span>, <span class="st">"Women"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">observed</span></span>
<span><span class="co">## Response: art (numeric)</span></span>
<span><span class="co">## Explanatory: fem (factor)</span></span>
<span><span class="co">## # A tibble: 1 × 1</span></span>
<span><span class="co">##    stat</span></span>
<span><span class="co">##   &lt;dbl&gt;</span></span>
<span><span class="co">## 1 0.412</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>From here, we compute a confidence interval for this mean by creating the bootstrap distribution via <code><a href="https://infer.tidymodels.org/reference/generate.html">generate()</a></code>; the same statistic is computed for each resampled version of the data:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">2101</span><span class="op">)</span></span>
<span><span class="va">bootstrapped</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">bioChemists</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/specify.html">specify</a></span><span class="op">(</span><span class="va">art</span> <span class="op">~</span> <span class="va">fem</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/generate.html">generate</a></span><span class="op">(</span>reps <span class="op">=</span> <span class="fl">2000</span>, type <span class="op">=</span> <span class="st">"bootstrap"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/calculate.html">calculate</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"diff in means"</span>, order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Men"</span>, <span class="st">"Women"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bootstrapped</span></span>
<span><span class="co">## Response: art (numeric)</span></span>
<span><span class="co">## Explanatory: fem (factor)</span></span>
<span><span class="co">## # A tibble: 2,000 × 2</span></span>
<span><span class="co">##   replicate  stat</span></span>
<span><span class="co">##       &lt;int&gt; &lt;dbl&gt;</span></span>
<span><span class="co">## 1         1 0.467</span></span>
<span><span class="co">## 2         2 0.107</span></span>
<span><span class="co">## 3         3 0.467</span></span>
<span><span class="co">## 4         4 0.308</span></span>
<span><span class="co">## 5         5 0.369</span></span>
<span><span class="co">## 6         6 0.428</span></span>
<span><span class="co">## # ℹ 1,994 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A percentile interval is calculated using:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">percentile_ci</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://infer.tidymodels.org/reference/get_confidence_interval.html">get_ci</a></span><span class="op">(</span><span class="va">bootstrapped</span><span class="op">)</span></span>
<span><span class="va">percentile_ci</span></span>
<span><span class="co">## # A tibble: 1 × 2</span></span>
<span><span class="co">##   lower_ci upper_ci</span></span>
<span><span class="co">##      &lt;dbl&gt;    &lt;dbl&gt;</span></span>
<span><span class="co">## 1    0.158    0.653</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <span class="pkg">infer</span> package has a high-level API for showing the analysis results, as shown in Figure @ref(fig:bootstrapped-mean).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://infer.tidymodels.org/reference/visualize.html">visualize</a></span><span class="op">(</span><span class="va">bootstrapped</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://infer.tidymodels.org/reference/shade_confidence_interval.html">shade_confidence_interval</a></span><span class="op">(</span>endpoints <span class="op">=</span> <span class="va">percentile_ci</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="figures/bootstrapped-mean-1.png" class="img-fluid figure-img" style="width:80.0%" alt="The bootstrap distribution of the difference in means. The highlighted region is the confidence interval, which does not include a value of zero."></p>
<figcaption class="figure-caption">The bootstrap distribution of the difference in means. The highlighted region is the confidence interval.</figcaption></figure>
</div>
</div>
</div>
<p>Since the interval visualized in in Figure @ref(fig:bootstrapped-mean) does not include zero, these results indicate that men have published more articles than women.</p>
<p>If we require a p-value, the <span class="pkg">infer</span> package can compute the value via a permutation test, shown in the following code. The syntax is very similar to the bootstrapping code we used earlier. We add a <code><a href="https://infer.tidymodels.org/reference/hypothesize.html">hypothesize()</a></code> verb to state the type of assumption to test and the <code><a href="https://infer.tidymodels.org/reference/generate.html">generate()</a></code> call contains an option to shuffle the data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">2102</span><span class="op">)</span></span>
<span><span class="va">permuted</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">bioChemists</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/specify.html">specify</a></span><span class="op">(</span><span class="va">art</span> <span class="op">~</span> <span class="va">fem</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/hypothesize.html">hypothesize</a></span><span class="op">(</span>null <span class="op">=</span> <span class="st">"independence"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/generate.html">generate</a></span><span class="op">(</span>reps <span class="op">=</span> <span class="fl">2000</span>, type <span class="op">=</span> <span class="st">"permute"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/calculate.html">calculate</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"diff in means"</span>, order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Men"</span>, <span class="st">"Women"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">permuted</span></span>
<span><span class="co">## Response: art (numeric)</span></span>
<span><span class="co">## Explanatory: fem (factor)</span></span>
<span><span class="co">## Null Hypothesis: independence</span></span>
<span><span class="co">## # A tibble: 2,000 × 2</span></span>
<span><span class="co">##   replicate     stat</span></span>
<span><span class="co">##       &lt;int&gt;    &lt;dbl&gt;</span></span>
<span><span class="co">## 1         1  0.201  </span></span>
<span><span class="co">## 2         2 -0.133  </span></span>
<span><span class="co">## 3         3  0.109  </span></span>
<span><span class="co">## 4         4 -0.195  </span></span>
<span><span class="co">## 5         5 -0.00128</span></span>
<span><span class="co">## 6         6 -0.102  </span></span>
<span><span class="co">## # ℹ 1,994 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The following visualization code is also very similar to the bootstrap approach. This code generates Figure @ref(fig:permutation-dist) where the vertical line signifies the observed value:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://infer.tidymodels.org/reference/visualize.html">visualize</a></span><span class="op">(</span><span class="va">permuted</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://infer.tidymodels.org/reference/shade_p_value.html">shade_p_value</a></span><span class="op">(</span>obs_stat <span class="op">=</span> <span class="va">observed</span>, direction <span class="op">=</span> <span class="st">"two-sided"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="figures/permutation-dist-1.png" class="img-fluid figure-img" style="width:80.0%" alt="The empirical distribution of the test statistic under the null hypothesis. The vertical line indicates the observed test statistic and is far away form the mainstream of the distribution."></p>
<figcaption class="figure-caption">Empirical distribution of the test statistic under the null hypothesis. The vertical line indicates the observed test statistic.</figcaption></figure>
</div>
</div>
</div>
<p>The actual p-value is:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">permuted</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://infer.tidymodels.org/reference/get_p_value.html">get_p_value</a></span><span class="op">(</span>obs_stat <span class="op">=</span> <span class="va">observed</span>, direction <span class="op">=</span> <span class="st">"two-sided"</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 1</span></span>
<span><span class="co">##   p_value</span></span>
<span><span class="co">##     &lt;dbl&gt;</span></span>
<span><span class="co">## 1   0.002</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The vertical line representing the null hypothesis in Figure @ref(fig:permutation-dist) is far away from the permutation distribution. This means, if in fact the null hypothesis were true, the likelihood is exceedingly small of observing data at least as extreme as what is at hand.</p>
<p>The two-sample tests shown in this section are probably suboptimal because they do not account for other factors that might explain the observed relationship between publication rate and sex. Let’s move to a more complex model that can consider additional covariates.</p>
</section><section id="log-linear-models" class="level2" data-number="21.3"><h2 data-number="21.3" class="anchored" data-anchor-id="log-linear-models">
<span class="header-section-number">21.3</span> Log-Linear Models</h2>
<p>The focus of the rest of this chapter will be on a generalized linear model <span class="citation" data-cites="Dobson99">(<a href="references.html#ref-Dobson99" role="doc-biblioref">Dobson 1999</a>)</span> where we assume the counts follow a Poisson distribution. For this model, the covariates/predictors enter the model in a log-linear fashion:</p>
<p><span class="math display">\[
\log(\lambda) = \beta_0 + \beta_1x_1 + \ldots + \beta_px_p
\]</span></p>
<p>where <span class="math inline">\(\lambda\)</span> is the expected value of the counts.</p>
<p>Let’s fit a simple model that contains all of the predictor columns. The <span class="pkg">poissonreg</span> package, a <span class="pkg">parsnip</span> extension package in tidymodels, will fit this model specification:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidymodels/poissonreg">poissonreg</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># default engine is 'glm'</span></span>
<span><span class="va">log_lin_spec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://parsnip.tidymodels.org/reference/poisson_reg.html">poisson_reg</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">log_lin_fit</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">log_lin_spec</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html">fit</a></span><span class="op">(</span><span class="va">art</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">bioChemists</span><span class="op">)</span></span>
<span><span class="va">log_lin_fit</span></span>
<span><span class="co">## parsnip model object</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Call:  stats::glm(formula = art ~ ., family = stats::poisson, data = data)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">## (Intercept)     femWomen   marMarried         kid5          phd         ment  </span></span>
<span><span class="co">##      0.3046      -0.2246       0.1552      -0.1849       0.0128       0.0255  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Degrees of Freedom: 914 Total (i.e. Null);  909 Residual</span></span>
<span><span class="co">## Null Deviance:       1820 </span></span>
<span><span class="co">## Residual Deviance: 1630  AIC: 3310</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://generics.r-lib.org/reference/tidy.html">tidy()</a></code> method succinctly summarizes the coefficients for the model (along with 90% confidence intervals):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="va">log_lin_fit</span>, conf.int <span class="op">=</span> <span class="cn">TRUE</span>, conf.level <span class="op">=</span> <span class="fl">0.90</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 7</span></span>
<span><span class="co">##   term        estimate std.error statistic  p.value conf.low conf.high</span></span>
<span><span class="co">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span><span class="co">## 1 (Intercept)   0.305    0.103       2.96  3.10e- 3   0.134     0.473 </span></span>
<span><span class="co">## 2 femWomen     -0.225    0.0546     -4.11  3.92e- 5  -0.315    -0.135 </span></span>
<span><span class="co">## 3 marMarried    0.155    0.0614      2.53  1.14e- 2   0.0545    0.256 </span></span>
<span><span class="co">## 4 kid5         -0.185    0.0401     -4.61  4.08e- 6  -0.251    -0.119 </span></span>
<span><span class="co">## 5 phd           0.0128   0.0264      0.486 6.27e- 1  -0.0305    0.0563</span></span>
<span><span class="co">## 6 ment          0.0255   0.00201    12.7   3.89e-37   0.0222    0.0288</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this output, the p-values correspond to separate hypothesis tests for each parameter:</p>
<span class="math display">\[\begin{align}
H_0&amp;: \beta_j = 0 \notag \\
H_a&amp;: \beta_j \ne 0 \notag
\end{align}\]</span>
<p>for each of the model parameters. Looking at these results, <code>phd</code> (the prestige of their department) may not have any relationship with the outcome.</p>
<p>While the Poisson distribution is the routine assumption for data like these, it may be beneficial to conduct a rough check of the model assumptions by fitting the models without using the Poisson likelihood to calculate the confidence intervals. The <span class="pkg">rsample</span> package has a convenience function to compute bootstrap confidence intervals for <code><a href="https://rdrr.io/r/stats/lm.html">lm()</a></code> and <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> models. We can use this function, while explicitly declaring <code>family = poisson</code>, to compute a large number of model fits. By default, we compute a 90% confidence bootstrap-t interval (percentile intervals are also available):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">2103</span><span class="op">)</span></span>
<span><span class="va">glm_boot</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">reg_intervals</span><span class="op">(</span><span class="va">art</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">bioChemists</span>, model_fn <span class="op">=</span> <span class="st">"glm"</span>, family <span class="op">=</span> <span class="va">poisson</span><span class="op">)</span></span>
<span><span class="va">glm_boot</span></span>
<span><span class="co">## # A tibble: 5 × 6</span></span>
<span><span class="co">##   term          .lower .estimate  .upper .alpha .method  </span></span>
<span><span class="co">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;    </span></span>
<span><span class="co">## 1 femWomen   -0.358      -0.226  -0.0856   0.05 student-t</span></span>
<span><span class="co">## 2 kid5       -0.298      -0.184  -0.0789   0.05 student-t</span></span>
<span><span class="co">## 3 marMarried  0.000264    0.155   0.317    0.05 student-t</span></span>
<span><span class="co">## 4 ment        0.0182      0.0256  0.0322   0.05 student-t</span></span>
<span><span class="co">## 5 phd        -0.0707      0.0130  0.102    0.05 student-t</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="rmdwarning">
<p>When we compare these results (in Figure @ref(fig:glm-intervals)) to the purely parametric results from <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code>, the bootstrap intervals are somewhat wider. If the data were truly Poisson, these intervals would have more similar widths.</p>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="figures/glm-intervals-1.png" class="img-fluid figure-img" alt="Two types of confidence intervals for the Poisson regression model. the interval for the PhD model is the only interval overlapping zero. The parametric intervals tend to be wider than the bootstrap intervals. " width="672"></p>
<figcaption class="figure-caption">Two types of confidence intervals for the Poisson regression model</figcaption></figure>
</div>
</div>
</div>
<p>Determining which predictors to include in the model is a difficult problem. One approach is to conduct likelihood ratio tests (LRT) <span class="citation" data-cites="McCullaghNelder89">(<a href="references.html#ref-McCullaghNelder89" role="doc-biblioref">McCullagh and Nelder 1989</a>)</span> between nested models. Based on the confidence intervals, we have evidence that a simpler model without <code>phd</code> may be sufficient. Let’s fit a smaller model, then conduct a statistical test:</p>
<span class="math display">\[\begin{align}
H_0&amp;: \beta_{phd} = 0 \notag \\
H_a&amp;: \beta_{phd} \ne 0 \notag
\end{align}\]</span>
<p>This hypothesis was previously tested when we showed the tidied results for <code>log_lin_fit</code>. That particular approach used results from a single model fit via a Wald statistic (i.e., the parameter divided by its standard error). For that approach, the p-value was 0.63. We can tidy the results for the LRT to get the p-value:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">log_lin_reduced</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">log_lin_spec</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html">fit</a></span><span class="op">(</span><span class="va">art</span> <span class="op">~</span> <span class="va">ment</span> <span class="op">+</span> <span class="va">kid5</span> <span class="op">+</span> <span class="va">fem</span> <span class="op">+</span> <span class="va">mar</span>, data <span class="op">=</span> <span class="va">bioChemists</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_fit_engine</a></span><span class="op">(</span><span class="va">log_lin_reduced</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_fit_engine</a></span><span class="op">(</span><span class="va">log_lin_fit</span><span class="op">)</span>,</span>
<span>  test <span class="op">=</span> <span class="st">"LRT"</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 2 × 6</span></span>
<span><span class="co">##   term                          df.residual residual.deviance    df deviance p.value</span></span>
<span><span class="co">##   &lt;chr&gt;                               &lt;dbl&gt;             &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">## 1 art ~ ment + kid5 + fem + mar         910             1635.    NA   NA      NA    </span></span>
<span><span class="co">## 2 art ~ fem + mar + kid5 + phd…         909             1634.     1    0.236   0.627</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The results are the same and, based on these and the confidence interval for this parameter, we’ll exclude <code>phd</code> from further analyses since it does not appear to be associated with the outcome.</p>
</section><section id="a-more-complex-model" class="level2" data-number="21.4"><h2 data-number="21.4" class="anchored" data-anchor-id="a-more-complex-model">
<span class="header-section-number">21.4</span> A More Complex Model</h2>
<p>We can move into even more complex models within our tidymodels approach. For count data, there are occasions where the number of zero counts is larger than what a simple Poisson distribution would prescribe. A more complex model appropriate for this situation is the zero-inflated Poisson (ZIP) model; see <span class="citation" data-cites="Mullahy">Mullahy (<a href="references.html#ref-Mullahy" role="doc-biblioref">1986</a>)</span>, <span class="citation" data-cites="Lambert1992">Lambert (<a href="references.html#ref-Lambert1992" role="doc-biblioref">1992</a>)</span>, and <span class="citation" data-cites="JSSv027i08">Zeileis, Kleiber, and Jackman (<a href="references.html#ref-JSSv027i08" role="doc-biblioref">2008</a>)</span>. Here, there are two sets of covariates: one for the count data and others that affect the probability (denoted as <span class="math inline">\(\pi\)</span>) of zeros. The equation for the mean <span class="math inline">\(\lambda\)</span> is:</p>
<p><span class="math display">\[\lambda = 0 \pi + (1 - \pi) \lambda_{nz}\]</span></p>
<p>where</p>
<span class="math display">\[\begin{align}
\log(\lambda_{nz}) &amp;= \beta_0 + \beta_1x_1 + \ldots + \beta_px_p \notag \\
\log\left(\frac{\pi}{1-\pi}\right) &amp;= \gamma_0 + \gamma_1z_1 + \ldots + \gamma_qz_q \notag
\end{align}\]</span>
<p>and the <span class="math inline">\(x\)</span> covariates affect the count values while the <span class="math inline">\(z\)</span> covariates influence the probability of a zero. The two sets of predictors do not need to be mutually exclusive.</p>
<p>We’ll fit a model with a full set of <span class="math inline">\(z\)</span> covariates:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">zero_inflated_spec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://parsnip.tidymodels.org/reference/poisson_reg.html">poisson_reg</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://parsnip.tidymodels.org/reference/set_engine.html">set_engine</a></span><span class="op">(</span><span class="st">"zeroinfl"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">zero_inflated_fit</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">zero_inflated_spec</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html">fit</a></span><span class="op">(</span><span class="va">art</span> <span class="op">~</span> <span class="va">fem</span> <span class="op">+</span> <span class="va">mar</span> <span class="op">+</span> <span class="va">kid5</span> <span class="op">+</span> <span class="va">ment</span> <span class="op">|</span> <span class="va">fem</span> <span class="op">+</span> <span class="va">mar</span> <span class="op">+</span> <span class="va">kid5</span> <span class="op">+</span> <span class="va">phd</span> <span class="op">+</span> <span class="va">ment</span>,</span>
<span>      data <span class="op">=</span> <span class="va">bioChemists</span><span class="op">)</span></span>
<span></span>
<span><span class="va">zero_inflated_fit</span></span>
<span><span class="co">## parsnip model object</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## pscl::zeroinfl(formula = art ~ fem + mar + kid5 + ment | fem + mar + kid5 + </span></span>
<span><span class="co">##     phd + ment, data = data)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Count model coefficients (poisson with log link):</span></span>
<span><span class="co">## (Intercept)     femWomen   marMarried         kid5         ment  </span></span>
<span><span class="co">##       0.621       -0.209        0.105       -0.143        0.018  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Zero-inflation model coefficients (binomial with logit link):</span></span>
<span><span class="co">## (Intercept)     femWomen   marMarried         kid5          phd         ment  </span></span>
<span><span class="co">##     -0.6086       0.1093      -0.3529       0.2195       0.0124      -0.1351</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since the coefficients for this model are also estimated using maximum likelihood, let’s try to use another likelihood ratio test to understand if the new model terms are helpful. We will <em>simultaneously</em> test that:</p>
<span class="math display">\[\begin{align}
H_0&amp;: \gamma_1 = 0, \gamma_2 = 0, \cdots, \gamma_5 = 0 \notag \\
H_a&amp;: \text{at least one } \gamma \ne 0  \notag
\end{align}\]</span>
<p>Let’s try ANOVA again:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_fit_engine</a></span><span class="op">(</span><span class="va">zero_inflated_fit</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_fit_engine</a></span><span class="op">(</span><span class="va">log_lin_reduced</span><span class="op">)</span>,</span>
<span>  test <span class="op">=</span> <span class="st">"LRT"</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Error in UseMethod("anova"): no applicable method for 'anova' applied to an object of class "zeroinfl"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>An <code><a href="https://rdrr.io/r/stats/anova.html">anova()</a></code> method isn’t implemented for <code>zeroinfl</code> objects!</p>
<p>An alternative is to use an <em>information criterion statistic</em>, such as the Akaike information criterion (AIC) <span class="citation" data-cites="claeskens2016statistical">(<a href="references.html#ref-claeskens2016statistical" role="doc-biblioref">Claeskens 2016</a>)</span>. This computes the log-likelihood (from the training set) and penalizes that value based on the training set size and the number of model parameters. In R’s parameterization, smaller AIC values are better. In this case, we are not conducting a formal statistical test but <em>estimating</em> the ability of the data to fit the data.</p>
<p>The results indicate that the ZIP model is preferable:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">zero_inflated_fit</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_fit_engine</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/AIC.html">AIC</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## [1] 3232</span></span>
<span><span class="va">log_lin_reduced</span>   <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_fit_engine</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/AIC.html">AIC</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## [1] 3312</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, it’s hard to contextualize this pair of single values and assess <em>how</em> different they actually are. To solve this problem, we’ll resample a large number of each of these two models. From these, we can compute the AIC values for each and determine how often the results favor the ZIP model. Basically, we will be characterizing the uncertainty of the AIC statistics to gauge their difference relative to the noise in the data.</p>
<p>We’ll also compute more bootstrap confidence intervals for the parameters in a bit so we specify the <code>apparent = TRUE</code> option when creating the bootstrap samples. This is required for some types of intervals.</p>
<p>First, we create the 4,000 model fits:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">zip_form</span> <span class="op">&lt;-</span> <span class="va">art</span> <span class="op">~</span> <span class="va">fem</span> <span class="op">+</span> <span class="va">mar</span> <span class="op">+</span> <span class="va">kid5</span> <span class="op">+</span> <span class="va">ment</span> <span class="op">|</span> <span class="va">fem</span> <span class="op">+</span> <span class="va">mar</span> <span class="op">+</span> <span class="va">kid5</span> <span class="op">+</span> <span class="va">phd</span> <span class="op">+</span> <span class="va">ment</span></span>
<span><span class="va">glm_form</span> <span class="op">&lt;-</span> <span class="va">art</span> <span class="op">~</span> <span class="va">fem</span> <span class="op">+</span> <span class="va">mar</span> <span class="op">+</span> <span class="va">kid5</span> <span class="op">+</span> <span class="va">ment</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">2104</span><span class="op">)</span></span>
<span><span class="va">bootstrap_models</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">bootstraps</span><span class="op">(</span><span class="va">bioChemists</span>, times <span class="op">=</span> <span class="fl">2000</span>, apparent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span></span>
<span>    glm <span class="op">=</span> <span class="fu">map</span><span class="op">(</span><span class="va">splits</span>, <span class="op">~</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html">fit</a></span><span class="op">(</span><span class="va">log_lin_spec</span>,       <span class="va">glm_form</span>, data <span class="op">=</span> <span class="fu">analysis</span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    zip <span class="op">=</span> <span class="fu">map</span><span class="op">(</span><span class="va">splits</span>, <span class="op">~</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html">fit</a></span><span class="op">(</span><span class="va">zero_inflated_spec</span>, <span class="va">zip_form</span>, data <span class="op">=</span> <span class="fu">analysis</span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">bootstrap_models</span></span>
<span><span class="co">## # Bootstrap sampling with apparent sample </span></span>
<span><span class="co">## # A tibble: 2,001 × 4</span></span>
<span><span class="co">##   splits            id            glm      zip     </span></span>
<span><span class="co">##   &lt;list&gt;            &lt;chr&gt;         &lt;list&gt;   &lt;list&gt;  </span></span>
<span><span class="co">## 1 &lt;split [915/355]&gt; Bootstrap0001 &lt;fit[+]&gt; &lt;fit[+]&gt;</span></span>
<span><span class="co">## 2 &lt;split [915/333]&gt; Bootstrap0002 &lt;fit[+]&gt; &lt;fit[+]&gt;</span></span>
<span><span class="co">## 3 &lt;split [915/337]&gt; Bootstrap0003 &lt;fit[+]&gt; &lt;fit[+]&gt;</span></span>
<span><span class="co">## 4 &lt;split [915/344]&gt; Bootstrap0004 &lt;fit[+]&gt; &lt;fit[+]&gt;</span></span>
<span><span class="co">## 5 &lt;split [915/351]&gt; Bootstrap0005 &lt;fit[+]&gt; &lt;fit[+]&gt;</span></span>
<span><span class="co">## 6 &lt;split [915/354]&gt; Bootstrap0006 &lt;fit[+]&gt; &lt;fit[+]&gt;</span></span>
<span><span class="co">## # ℹ 1,995 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can extract the model fits and their corresponding AIC values:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">bootstrap_models</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">bootstrap_models</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span></span>
<span>    glm_aic <span class="op">=</span> <span class="fu">map_dbl</span><span class="op">(</span><span class="va">glm</span>, <span class="op">~</span> <span class="fu"><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_fit_engine</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/AIC.html">AIC</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    zip_aic <span class="op">=</span> <span class="fu">map_dbl</span><span class="op">(</span><span class="va">zip</span>, <span class="op">~</span> <span class="fu"><a href="https://hardhat.tidymodels.org/reference/hardhat-extract.html">extract_fit_engine</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/AIC.html">AIC</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">bootstrap_models</span><span class="op">$</span><span class="va">zip_aic</span> <span class="op">&lt;</span> <span class="va">bootstrap_models</span><span class="op">$</span><span class="va">glm_aic</span><span class="op">)</span></span>
<span><span class="co">## [1] 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It seems definitive from these results that accounting for the excessive number of zero counts is a good idea.</p>
<div class="rmdnote">
<p>We could have used <code>fit_resamples()</code> or a workflow set to conduct these computations. In this section, we used <code>mutate()</code> and <code>map()</code> to compute the models to demonstrate how one might use tidymodels tools for models that are not supported by one of the <span class="pkg">parsnip</span> packages.</p>
</div>
<p>Since we have computed the resampled model fits, let’s create bootstrap intervals for the zero probability model coefficients (i.e., the <span class="math inline">\(\gamma_j\)</span>). We can extract these with the <code><a href="https://generics.r-lib.org/reference/tidy.html">tidy()</a></code> method and use the <code>type = "zero"</code> option to obtain these estimates:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">bootstrap_models</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">bootstrap_models</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>zero_coefs  <span class="op">=</span> <span class="fu">map</span><span class="op">(</span><span class="va">zip</span>, <span class="op">~</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="va">.x</span>, type <span class="op">=</span> <span class="st">"zero"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># One example:</span></span>
<span><span class="va">bootstrap_models</span><span class="op">$</span><span class="va">zero_coefs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## # A tibble: 6 × 6</span></span>
<span><span class="co">##   term        type  estimate std.error statistic   p.value</span></span>
<span><span class="co">##   &lt;chr&gt;       &lt;chr&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span><span class="co">## 1 (Intercept) zero   -0.128     0.497     -0.257 0.797    </span></span>
<span><span class="co">## 2 femWomen    zero   -0.0764    0.319     -0.240 0.811    </span></span>
<span><span class="co">## 3 marMarried  zero   -0.112     0.365     -0.307 0.759    </span></span>
<span><span class="co">## 4 kid5        zero    0.270     0.186      1.45  0.147    </span></span>
<span><span class="co">## 5 phd         zero   -0.178     0.132     -1.35  0.177    </span></span>
<span><span class="co">## 6 ment        zero   -0.123     0.0315    -3.91  0.0000935</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It’s a good idea to visualize the bootstrap distributions of the coefficients, as in Figure @ref(fig:zip-bootstrap).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">bootstrap_models</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">unnest</span><span class="op">(</span><span class="va">zero_coefs</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">25</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">term</span>, scales <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"gray70"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="figures/zip-bootstrap-1.png" class="img-fluid figure-img" alt="Bootstrap distributions of the ZIP model coefficients. The vertical lines indicate the observed estimates. The ment predictor that appears to be important to the model." width="672"></p>
<figcaption class="figure-caption">Bootstrap distributions of the ZIP model coefficients. The vertical lines indicate the observed estimates.</figcaption></figure>
</div>
</div>
</div>
<p>One of the covariates (<code>ment</code>) that appears to be important has a very skewed distribution. The extra space in some of the facets indicates there are some outliers in the estimates. This <em>might</em> occur when models did not converge; those results probably should be excluded from the resamples. For the results visualized in Figure @ref(fig:zip-bootstrap), the outliers are due only to extreme parameter estimates; all of the models converged.</p>
<p>The <span class="pkg">rsample</span> package contains a set of functions named <code>int_*()</code> that compute different types of bootstrap intervals. Since the <code><a href="https://generics.r-lib.org/reference/tidy.html">tidy()</a></code> method contains standard error estimates, the bootstrap-t intervals can be computed. We’ll also compute the standard percentile intervals. By default, 90% confidence intervals are computed.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">bootstrap_models</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">int_pctl</span><span class="op">(</span><span class="va">zero_coefs</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 6</span></span>
<span><span class="co">##   term        .lower .estimate  .upper .alpha .method   </span></span>
<span><span class="co">##   &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;     </span></span>
<span><span class="co">## 1 (Intercept) -1.75    -0.621   0.423    0.05 percentile</span></span>
<span><span class="co">## 2 femWomen    -0.521    0.115   0.818    0.05 percentile</span></span>
<span><span class="co">## 3 kid5        -0.327    0.218   0.677    0.05 percentile</span></span>
<span><span class="co">## 4 marMarried  -1.20    -0.381   0.362    0.05 percentile</span></span>
<span><span class="co">## 5 ment        -0.401   -0.162  -0.0513   0.05 percentile</span></span>
<span><span class="co">## 6 phd         -0.276    0.0220  0.327    0.05 percentile</span></span>
<span><span class="va">bootstrap_models</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">int_t</span><span class="op">(</span><span class="va">zero_coefs</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 6</span></span>
<span><span class="co">##   term        .lower .estimate  .upper .alpha .method  </span></span>
<span><span class="co">##   &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;    </span></span>
<span><span class="co">## 1 (Intercept) -1.61    -0.621   0.321    0.05 student-t</span></span>
<span><span class="co">## 2 femWomen    -0.482    0.115   0.671    0.05 student-t</span></span>
<span><span class="co">## 3 kid5        -0.211    0.218   0.599    0.05 student-t</span></span>
<span><span class="co">## 4 marMarried  -0.988   -0.381   0.290    0.05 student-t</span></span>
<span><span class="co">## 5 ment        -0.324   -0.162  -0.0275   0.05 student-t</span></span>
<span><span class="co">## 6 phd         -0.274    0.0220  0.291    0.05 student-t</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>From these results, we can get a good idea of which predictor(s) to include in the zero count probability model. It may be sensible to refit a smaller model to assess if the bootstrap distribution for <code>ment</code> is still skewed.</p>
</section><section id="inference-options" class="level2" data-number="21.5"><h2 data-number="21.5" class="anchored" data-anchor-id="inference-options">
<span class="header-section-number">21.5</span> More Inferential Analysis</h2>
<p>This chapter demonstrated just a small subset of what is available for inferential analysis in tidymodels and has focused on resampling and frequentist methods. Arguably, Bayesian analysis is a very effective and often superior approach for inference. A variety of Bayesian models are available via <span class="pkg">parsnip</span>. Additionally, the <span class="pkg">multilevelmod</span> package enables users to fit hierarchical Bayesian and non-Bayesian models (e.g., mixed models). The <span class="pkg">broom.mixed</span> and <span class="pkg">tidybayes</span> packages are excellent tools for extracting data for plots and summaries. Finally, for data sets with a single hierarchy, such as simple longitudinal or repeated measures data, <span class="pkg">rsample</span>’s <code>group_vfold_cv()</code> function facilitates straightforward out-of-sample characterizations of model performance.</p>
</section><section id="inference-summary" class="level2" data-number="21.6"><h2 data-number="21.6" class="anchored" data-anchor-id="inference-summary">
<span class="header-section-number">21.6</span> Chapter Summary</h2>
<p>The tidymodels framework is for more than predictive modeling alone. Packages and functions from tidymodels can be used for hypothesis testing, as well as fitting and assessing inferential models. The tidymodels framework provides support for working with non-tidymodels R models, and can help assess the statistical qualities of your models.</p>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-claeskens2016statistical" class="csl-entry" role="listitem">
Claeskens, G. 2016. <span>“Statistical Model Choice.”</span> <em>Annual Review of Statistics and Its Application</em> 3: 233–56.
</div>
<div id="ref-davison1997bootstrap" class="csl-entry" role="listitem">
Davison, A, and D Hinkley. 1997. <em>Bootstrap Methods and Their Application</em>. Vol. 1. Cambridge university press.
</div>
<div id="ref-Dobson99" class="csl-entry" role="listitem">
Dobson, A. 1999. <em>An Introduction to Generalized Linear Models</em>. Chapman; Hall: Boca Raton.
</div>
<div id="ref-ModernDive" class="csl-entry" role="listitem">
Ismay, C, and A Kim. 2021. <em>Statistical Inference via Data Science: A ModernDive into r and the Tidyverse</em>. Chapman; Hall/CRC. <a href="https://moderndive.com/">https://moderndive.com/</a>.
</div>
<div id="ref-Lambert1992" class="csl-entry" role="listitem">
Lambert, D. 1992. <span>“Zero-Inflated Poisson Regression, with an Application to Defects in Manufacturing.”</span> <em>Technometrics</em> 34 (1): 1–14.
</div>
<div id="ref-Long1992" class="csl-entry" role="listitem">
Long, J. 1992. <span>“<span class="nocase">Measures of Sex Differences in Scientific Productivity*</span>.”</span> <em>Social Forces</em> 71 (1): 159–78.
</div>
<div id="ref-McCullaghNelder89" class="csl-entry" role="listitem">
McCullagh, P, and J Nelder. 1989. <em>Generalized Linear Models</em>. London: Chapman; Hall.
</div>
<div id="ref-Mullahy" class="csl-entry" role="listitem">
Mullahy, J. 1986. <span>“Specification and Testing of Some Modified Count Data Models.”</span> <em>Journal of Econometrics</em> 33 (3): 341–65.
</div>
<div id="ref-pvalue" class="csl-entry" role="listitem">
Wasserstein, R, and N Lazar. 2016. <span>“The <span>ASA</span> Statement on p-Values: Context, Process, and Purpose.”</span> <em>The American Statistician</em> 70 (2): 129–33.
</div>
<div id="ref-JSSv027i08" class="csl-entry" role="listitem">
Zeileis, A, C Kleiber, and S Jackman. 2008. <span>“Regression Models for Count Data in <span>R</span>.”</span> <em>Journal of Statistical Software</em> 27 (8): 1–25. <a href="https://www.jstatsoft.org/v027/i08">https://www.jstatsoft.org/v027/i08</a>.
</div>
</div>
</section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><hr>
<ol>
<li id="fn1"><p>The <code>T</code> argument allows us to account for the time when the events (publications) were counted, which was three years for both men and women. There are more men than women in these data, but <code><a href="https://rdrr.io/r/stats/poisson.test.html">poisson.test()</a></code> has limited functionality so more sophisticated analysis can be used to account for this difference.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./20-ensemble-models.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Ensembles of Models</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./pre-proc-table.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Recommended Preprocessing</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Modelado Ordenado con R fue escrito por Max Kuhn, y Julia Silge.</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Este libro fue creado usando <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>


</body></html>